<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Copy Number Alteration (CNA) Calling from muscadet object — cnaCalling • muscadet</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Copy Number Alteration (CNA) Calling from muscadet object — cnaCalling"><meta name="description" content="Performs copy number alteration (CNA) analysis on a muscadet
object by processing allelic and coverage counts across clusters and
evaluating cell fractions and copy numbers."><meta property="og:description" content="Performs copy number alteration (CNA) analysis on a muscadet
object by processing allelic and coverage counts across clusters and
evaluating cell fractions and copy numbers."><meta property="og:image" content="https://icagen.github.io/muscadet/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">muscadet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/muscadet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ICAGEN/muscadet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Copy Number Alteration (CNA) Calling from muscadet object</h1>
      <small class="dont-index">Source: <a href="https://github.com/ICAGEN/muscadet/blob/main/R/cnaCalling.R" class="external-link"><code>R/cnaCalling.R</code></a></small>
      <div class="d-none name"><code>cnaCalling.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Performs copy number alteration (CNA) analysis on a <code><a href="muscadet-class.html">muscadet</a></code>
object by processing allelic and coverage counts across clusters and
evaluating cell fractions and copy numbers.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cnaCalling</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  omics.coverage <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  depthmin.a.clusters <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  depthmin.c.clusters <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  depthmin.a.allcells <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  depthmin.c.allcells <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  depthmin.nor <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  depthmax.nor <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  het.thresh <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  snp.nbhd <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  hetscale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cval1 <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  cval2 <span class="op">=</span> <span class="fl">150</span>,</span>
<span>  min.nhet <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  clonal.thresh <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  dist.breakpoints <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  ploidy <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p>This function uses several functions from the <a href="https://rdrr.io/pkg/facets/man/facets-package.html" class="external-link">facets-package</a> package,
including: <code><a href="https://rdrr.io/pkg/facets/man/facets-internal.html" class="external-link">facets::clustersegs()</a></code>, <code><a href="https://rdrr.io/pkg/facets/man/emcncf.html" class="external-link">facets::emcncf()</a></code>,
<code><a href="https://rdrr.io/pkg/facets/man/facets-internal.html" class="external-link">facets::findDiploidLogR()</a></code>, <code><a href="https://rdrr.io/pkg/facets/man/fitcncf.html" class="external-link">facets::fitcncf()</a></code>, <code><a href="https://rdrr.io/pkg/facets/man/procSample.html" class="external-link">facets::procSample()</a></code>,
<code><a href="https://rdrr.io/pkg/facets/man/facets-internal.html" class="external-link">facets::procSnps()</a></code>, and adapted function <code><a href="preProcSample2.html">preProcSample2()</a></code>.</p>
<p>Seshan VE, Shen R (2021). <em>facets: Cellular Fraction and Copy Numbers from
Tumor Sequencing</em>. R package version 0.6.2,
<a href="https://github.com/mskcc/facets" class="external-link">https://github.com/mskcc/facets</a>.</p>
    </div>
    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A <code><a href="muscadet-class.html">muscadet</a></code> object. Must contain:</p><ul><li><p>Clustering assignments in the <code>cnacalling$clusters</code> slot
(use <code><a href="assignClusters.html">assignClusters()</a></code>).</p></li>
<li><p>Combined allelic and coverage counts per cluster in the
<code>cnacalling$combined.counts</code> slot (use <code><a href="mergeCounts.html">mergeCounts()</a></code>).</p></li>
</ul></dd>


<dt id="arg-omics-coverage">omics.coverage<a class="anchor" aria-label="anchor" href="#arg-omics-coverage"></a></dt>
<dd><p>A vector of omics names to select for coverage log R
ratio data. RECOMMENDED: select "ATAC" when ATAC and RNA omics are
available, the ATAC coverage (DNA) signal is less noisy than RNA signal. By
default, <code>NULL</code> selects all available data.</p></dd>


<dt id="arg-depthmin-a-clusters">depthmin.a.clusters<a class="anchor" aria-label="anchor" href="#arg-depthmin-a-clusters"></a></dt>
<dd><p>Minimum allelic depth per clusters in tumor cells
(default: 30).</p></dd>


<dt id="arg-depthmin-c-clusters">depthmin.c.clusters<a class="anchor" aria-label="anchor" href="#arg-depthmin-c-clusters"></a></dt>
<dd><p>Minimum coverage depth per clusters in tumor cells
(default: 50).</p></dd>


<dt id="arg-depthmin-a-allcells">depthmin.a.allcells<a class="anchor" aria-label="anchor" href="#arg-depthmin-a-allcells"></a></dt>
<dd><p>Minimum allelic depth for all tumor cells
(default: 30).</p></dd>


<dt id="arg-depthmin-c-allcells">depthmin.c.allcells<a class="anchor" aria-label="anchor" href="#arg-depthmin-c-allcells"></a></dt>
<dd><p>Minimum coverage depth for all tumor cells
(default: 50).</p></dd>


<dt id="arg-depthmin-nor">depthmin.nor<a class="anchor" aria-label="anchor" href="#arg-depthmin-nor"></a></dt>
<dd><p>Minimum coverage depth for normal sample (default: 0).</p></dd>


<dt id="arg-depthmax-nor">depthmax.nor<a class="anchor" aria-label="anchor" href="#arg-depthmax-nor"></a></dt>
<dd><p>Maximum coverage depth for normal sample (default: 1000).</p></dd>


<dt id="arg-het-thresh">het.thresh<a class="anchor" aria-label="anchor" href="#arg-het-thresh"></a></dt>
<dd><p>VAF (Variant Allele Frequency) threshold to call variant
positions heterozygous for <code><a href="preProcSample2.html">preProcSample2()</a></code> (default: 0.25).</p></dd>


<dt id="arg-snp-nbhd">snp.nbhd<a class="anchor" aria-label="anchor" href="#arg-snp-nbhd"></a></dt>
<dd><p>Window size for selecting SNP loci to reduce serial
correlation for <code><a href="preProcSample2.html">preProcSample2()</a></code> (default: 250).</p></dd>


<dt id="arg-hetscale">hetscale<a class="anchor" aria-label="anchor" href="#arg-hetscale"></a></dt>
<dd><p>Logical value indicating whether log odds ratio (logOR)
should be scaled to give more weight in the test statistics for
segmentation and clustering <code><a href="preProcSample2.html">preProcSample2()</a></code>. (default: <code>TRUE</code>)</p></dd>


<dt id="arg-cval-">cval1<a class="anchor" aria-label="anchor" href="#arg-cval-"></a></dt>
<dd><p>Critical value for segmentation for <code><a href="preProcSample2.html">preProcSample2()</a></code> (default:
25).</p></dd>


<dt id="arg-cval-">cval2<a class="anchor" aria-label="anchor" href="#arg-cval-"></a></dt>
<dd><p>Critical value for segmentation for <code><a href="https://rdrr.io/pkg/facets/man/procSample.html" class="external-link">facets::procSample()</a></code>
(default: 150).</p></dd>


<dt id="arg-min-nhet">min.nhet<a class="anchor" aria-label="anchor" href="#arg-min-nhet"></a></dt>
<dd><p>Minimum number of heterozygous positions in a segment for
<code><a href="https://rdrr.io/pkg/facets/man/procSample.html" class="external-link">facets::procSample()</a></code> and <code><a href="https://rdrr.io/pkg/facets/man/emcncf.html" class="external-link">facets::emcncf()</a></code> (default: 5).</p></dd>


<dt id="arg-clonal-thresh">clonal.thresh<a class="anchor" aria-label="anchor" href="#arg-clonal-thresh"></a></dt>
<dd><p>Threshold of minimum cell proportion to label a
segment as clonal (default: 0.9).</p></dd>


<dt id="arg-dist-breakpoints">dist.breakpoints<a class="anchor" aria-label="anchor" href="#arg-dist-breakpoints"></a></dt>
<dd><p>Minimum distance between breakpoints to define
distinct segments (default: 1e6).</p></dd>


<dt id="arg-ploidy">ploidy<a class="anchor" aria-label="anchor" href="#arg-ploidy"></a></dt>
<dd><p>Specifies ploidy assumption: <code>"auto"</code>, <code>"median"</code>, or numeric
value (default: <code>"auto"</code>).</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, suppresses informative messages during
execution. Default is <code>FALSE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A modified <code><a href="muscadet-class.html">muscadet</a></code> object with added CNA analysis results in
the <code>cnacalling</code> slot, including: filtered counts and positions, segmentation
data for clusters and all cells, consensus segments across clusters based on
breakpoints, diploid log R ratio, purity and ploidy.</p>
<p>Details of the cnacalling slot:</p><ul><li><p><code>combined.counts.filtered</code>: Filtered counts per clusters.</p></li>
<li><p><code>combined.counts.allcells</code>: Counts summed for all cells (no
cluster distinction).</p></li>
<li><p><code>combined.counts.allcells.filtered</code>: Filtered counts summed for
all cells (no cluster distinction).</p></li>
<li><p><code>positions</code>: Data frame of positions from the per cluster
analysis. Positions in rows and associated data in columns: <code>chrom</code>,
<code>maploc</code> (position), <code>rCountT</code> (read count in tumor), <code>rCountN</code> (read count
in normal), <code>vafT</code> (variant allele frequency in tumor), <code>vafN</code> (variant
allele frequency in normal), <code>cluster</code> (cluster id), <code>signal</code> (whether the
counts come from coverage or allelic data), <code>het</code> (heterozygous status),
<code>keep</code> (whether to keep position), <code>gcpct</code> (GC percentage), <code>gcbias</code> (GC
bias correction), <code>cnlr</code> (log R ratio), <code>valor</code> (log odds ratio), <code>lorvar</code>
(variance of log odds ratio), <code>seg0</code>, <code>seg_ori</code> (segment original id within
each cluster), <code>seg</code> (segment id), <code>segclust</code> (cluster of segments id),
<code>vafT.allcells</code> (vairiant allele frequency in all tumor cells), <code>colVAR</code>
(integer for allelic position color depending on <code>vafT.allcells</code>).</p></li>
<li><p><code>segments</code>: Data frame of segments from the per cluster
analysis. Segments in rows and associated data in columns: <code>chrom</code>, <code>seg</code>
(segment id), <code>num.mark</code> (number of positions in segment), <code>nhet</code> (number
of heterezygous positions in segment), <code>cnlr.median</code> (segment log R ratio
median), <code>mafR</code> (segment square of expected log odds ratio), vafT.median
(segment variant allele frequency median), <code>cluster</code> (cluster id),
<code>seg_ori</code> (segment original id within each cluster), <code>segclust</code> (cluster of
segments id), <code>cnlr.median.clust</code> (segment cluster log R ratio median),
<code>mafR.clust</code> (segment cluster square of expected log odds ratio), <code>cf</code>
(cell fraction), <code>tcn</code> (total copy number), <code>lcn</code> (lower copy number),
<code>start</code>, <code>end</code>, <code>cf.em</code> (cell fraction computed with EM algorithm),
<code>tcn.em</code>, (total copy number computed with EM algorithm), <code>lcn.em</code> (lower
copy number computed with EM algorithm).</p></li>
<li><p><code>positions.allcells</code>: Same as <code>positions</code> but from the all cells analysis.</p></li>
<li><p><code>segments.allcells</code>: Same as <code>segments</code> but from the all cells analysis.</p></li>
<li><p><code>consensus.segs</code>: Data frame of unique consensus segments across
clusters, with the <code>cna</code> (<code>logical</code>) and <code>cna_clonal</code> (<code>logical</code>)
information.</p></li>
<li><p><code>table</code>: Data frame of consensus segments across clusters with
associated information per cluster in columns: <code>chrom</code>, <code>start</code>, <code>end</code>, <code>id</code>,
<code>cluster</code>, <code>cf.em</code> (cell fraction computed with EM algorithm), <code>tcn.em</code>
(total copy number computed with EM algorithm), <code>lcn.em</code> (lower copy number
computed with EM algorithm), <code>ncells</code> (number of cells in cluster),
<code>prop.cluster</code> (proportion of cells per cluster), <code>gnl</code> (gain;neutral;loss :
1;0;-1), <code>loh</code> (loss of heterozygosity status), <code>state</code> (state of segments),
<code>cna</code> (whether the segment is a CNA), <code>cna_state</code> (state of CNA segments),
<code>prop.tot</code> (proportion of cells with the same state per segment),
<code>state_clonal</code> (state of the segment if its <code>prop.tot</code> is above
<code>clonal.thresh</code>), <code>cna_clonal</code> (whether the segment is a clonal CNA),
<code>cna_clonal_state</code> (state of clonal CNA segments).</p></li>
<li><p><code>ncells</code>: Vector of number of cells per cluster.</p></li>
<li><p><code>dipLogR.clusters</code>: Diploid log R ratio from the per cluster analysis.</p></li>
<li><p><code>dipLogR.allcells</code>: Diploid log R ratio from the all cells analysis.</p></li>
<li><p><code>purity.clusters</code>: Purity from the per cluster analysis.</p></li>
<li><p><code>purity.allcells</code>: Purity from the all cells analysis.</p></li>
<li><p><code>ploidy.clusters</code>: Ploidy from the per cluster analysis.</p></li>
<li><p><code>ploidy.allcells</code>: Ploidy from the all cells analysis.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Shen R, Seshan VE. FACETS: allele-specific copy number and clonal
heterogeneity analysis tool for high-throughput DNA sequencing. Nucleic
Acids Res. 2016 Sep 19;44(16):e131. doi:
<a href="http://doi.org/10.1093/nar/gkw520" class="external-link">10.1093/nar/gkw520</a>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="assignClusters.html">assignClusters()</a></code>, <code><a href="mergeCounts.html">mergeCounts()</a></code>,
<code><a href="preProcSample2.html">preProcSample2()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"facets"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: pctGCdata</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Load example muscadet object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">muscadet_obj</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">muscadet_obj</span> <span class="op">&lt;-</span> <span class="fu">cnaCalling</span><span class="op">(</span><span class="va">muscadet_obj</span>,</span></span>
<span class="r-in"><span>                           omics.coverage <span class="op">=</span> <span class="st">"ATAC"</span>,</span></span>
<span class="r-in"><span>                           depthmin.a.clusters <span class="op">=</span> <span class="fl">3</span>, <span class="co"># set low thresholds for example data</span></span></span>
<span class="r-in"><span>                           depthmin.c.clusters <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                           depthmin.a.allcells <span class="op">=</span> <span class="fl">3</span>,</span></span>
<span class="r-in"><span>                           depthmin.c.allcells <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                           depthmin.nor <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selecting coverage data from omic(s): ATAC</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering positions per clusters based on provided filters...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing segmentation per cluster...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finding diploid log R ratio on clusters...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Diploid log R ratio = -0.566520511224907</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing cell fractions and copy numbers on clusters...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering positions on all cells based on provided filters...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing segmentation on all cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing cell fractions and copy numbers on all cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finding consensus segments between clusters...</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marie Denoulet, Nils Giordano, Eric Letouzé, Céline Vallot, Stéphane Minvielle.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

