<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute log R ratios for scRNA-seq data — computeLogRatioRNA • muscadet</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute log R ratios for scRNA-seq data — computeLogRatioRNA"><meta name="description" content="Compute log R ratios from raw count matrices with method specifically adapted
for scRNA-seq data."><meta property="og:description" content="Compute log R ratios from raw count matrices with method specifically adapted
for scRNA-seq data."><meta property="og:image" content="https://icagen.github.io/muscadet/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">muscadet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/muscadet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ICAGEN/muscadet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Compute log R ratios for scRNA-seq data</h1>
      <small class="dont-index">Source: <a href="https://github.com/ICAGEN/muscadet/blob/main/R/computeLogRatio.R" class="external-link"><code>R/computeLogRatio.R</code></a></small>
      <div class="d-none name"><code>computeLogRatioRNA.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Compute log R ratios from raw count matrices with method specifically adapted
for scRNA-seq data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">computeLogRatioRNA</span><span class="op">(</span></span>
<span>  <span class="va">matTumor</span>,</span>
<span>  <span class="va">matRef</span>,</span>
<span>  <span class="va">genesCoord</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  genesPerWindow <span class="op">=</span> <span class="fl">101</span>,</span>
<span>  refReads <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  refMeanReads <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  thresh_capping <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  all_steps <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-mattumor">matTumor<a class="anchor" aria-label="anchor" href="#arg-mattumor"></a></dt>
<dd><p>Raw count matrix <em>features x cells</em> for tumor/sample cells
(<code>matrix</code> or <code>dgCMatrix</code>).</p></dd>


<dt id="arg-matref">matRef<a class="anchor" aria-label="anchor" href="#arg-matref"></a></dt>
<dd><p>Raw count matrix <em>features x cells</em> for reference cells
(<code>matrix</code> or <code>dgCMatrix</code>).</p></dd>


<dt id="arg-genescoord">genesCoord<a class="anchor" aria-label="anchor" href="#arg-genescoord"></a></dt>
<dd><p>Data frame of gene coordinates with columns <code>CHROM</code>,
<code>start</code>, <code>end</code>, <code>id</code> (<code>data.frame</code>).</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>Reference genome name among: "hg38", "hg19" and "mm10"
(<code>character</code>). By default: "hg38".</p></dd>


<dt id="arg-genesperwindow">genesPerWindow<a class="anchor" aria-label="anchor" href="#arg-genesperwindow"></a></dt>
<dd><p>Number of genes per moving window (<code>integer</code> value). By
default: <code>101</code>.</p></dd>


<dt id="arg-refreads">refReads<a class="anchor" aria-label="anchor" href="#arg-refreads"></a></dt>
<dd><p>Minimum of reads in reference cells (<code>integer</code> value). By
default: <code>100</code>.</p></dd>


<dt id="arg-refmeanreads">refMeanReads<a class="anchor" aria-label="anchor" href="#arg-refmeanreads"></a></dt>
<dd><p>Minimum of average reads per reference cell (<code>integer</code>
value). By default: <code>0.01</code>.</p></dd>


<dt id="arg-thresh-capping">thresh_capping<a class="anchor" aria-label="anchor" href="#arg-thresh-capping"></a></dt>
<dd><p>Threshold to cap the range of log R ratio values
(<code>numeric</code> value). By default: <code>3</code>.</p></dd>


<dt id="arg-all-steps">all_steps<a class="anchor" aria-label="anchor" href="#arg-all-steps"></a></dt>
<dd><p><code>TRUE</code> or <code>FALSE</code> (<code>logical</code>). Whether to keep intermediate
result from every step in the final object. By default: <code>FALSE</code>.</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, suppresses informative messages during
execution. Default is <code>FALSE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>If <code>all_steps</code> is set to <code>FALSE</code>, a list containing:</p><dl><dt><code>matTumor</code></dt>
<dd><p>Matrix of log R ratio values <em>features x cells</em> for the
tumor/sample cells (<code>matrix</code>).</p></dd>

<dt><code>matRef</code></dt>
<dd><p>Matrix of log R ratio values <em>features x cells</em> for the
reference cells (<code>matrix</code>).</p></dd>

<dt><code>params</code></dt>
<dd><p>List of parameters set for the <code>genesPerWindow</code>, <code>refMeans</code>,
<code>refMeanReads</code> and <code>thresh_capping</code> arguments (<code>list</code>).</p></dd>

<dt><code>coord</code></dt>
<dd><p>Data frame of coordinates for windows of peaks and associated
data along the different steps (<code>data.frame</code>).
Columns :</p><ul><li><p><code>CHROM</code>, <code>start</code>, <code>end</code>, <code>id</code>: coordinates and name of genes.</p></li>
<li><p><code>sumReads.tum/ref</code>: sum of read counts for all cells in tumor or reference
cells.</p></li>
<li><p><code>meanReads.tum/ref</code>: mean of read counts per cells for tumor or reference
cells.</p></li>
<li><p><code>sdReads.tum/ref</code>: standard deviation of read counts per cells for tumor
or reference cells.</p></li>
<li><p><code>keep</code>: logical, <code>TRUE</code> for genes to keep after filtering based on
reference coverage (depends on <code>refReads</code> and <code>refMeanReads</code> arguments).</p></li>
<li><p><code>meanReads/sdReads.norm.tum/ref</code>: mean/sd of normalized counts per million
for tumor/reference cells.</p></li>
<li><p><code>meanLRR/sdReads.raw.tum/ref</code>: mean/sd of raw log R ratio (LRR) for
tumor/reference cells.</p></li>
<li><p><code>meanLRR/sdLRR.cap.tum/ref</code>: mean/sd of capped log R ratio (LRR) for
tumor/reference cells (depends on <code>thresh_capping</code> argument).</p></li>
<li><p><code>meanLRR/sdLRR.smoo.tum/ref</code>: mean/sd of smoothed log R ratio (LRR) for
tumor/reference cells (means of moving windows defined by the
<code>genesPerWindow</code> argument).</p></li>
<li><p><code>meanLRR/sdLRR.cent.tum/ref</code>: mean/sd of centered log R ratio (LRR) for
tumor/reference cells.</p></li>
<li><p><code>meanLRR/sdLRR.corr.tum/ref</code>: mean/sd of final log R ratio (LRR) corrected
by reference variability for tumor/reference cells.</p></li>
</ul></dd>

If all_steps is set to TRUE, the previously described list can be found
for each step in a list element.

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The raw count matrix is transformed into log R ratios through the following
steps:</p><ul><li><p>Match genes in count matrix with coordinates</p></li>
<li><p>Filtering on coverage (<code>refReads</code> and <code>refMeanReads</code> arguments)</p></li>
<li><p>Normalization for sequencing depth</p></li>
<li><p>Log transformation and normalization by reference data: log R ratio</p></li>
<li><p>Capping the range of values (<code>thresh_capping</code> argument)</p></li>
<li><p>Smoothing on genes windows</p></li>
<li><p>Centering of cells</p></li>
<li><p>Correcting by reference variability</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other computeLogRatio:
<code><a href="computeLogRatio.html">computeLogRatio</a>()</code>,
<code><a href="computeLogRatioATAC.html">computeLogRatioATAC</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Create muscomic objects</span></span></span>
<span class="r-in"><span><span class="va">atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  type <span class="op">=</span> <span class="st">"ATAC"</span>,</span></span>
<span class="r-in"><span>  mat_counts <span class="op">=</span> <span class="va">mat_counts_atac_tumor</span>,</span></span>
<span class="r-in"><span>  allele_counts <span class="op">=</span> <span class="va">allele_counts_atac_tumor</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="va">peaks</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  type <span class="op">=</span> <span class="st">"RNA"</span>,</span></span>
<span class="r-in"><span>  mat_counts <span class="op">=</span> <span class="va">mat_counts_rna_tumor</span>,</span></span>
<span class="r-in"><span>  allele_counts <span class="op">=</span> <span class="va">allele_counts_rna_tumor</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="va">genes</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">atac_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  type <span class="op">=</span> <span class="st">"ATAC"</span>,</span></span>
<span class="r-in"><span>  mat_counts <span class="op">=</span> <span class="va">mat_counts_atac_ref</span>,</span></span>
<span class="r-in"><span>  allele_counts <span class="op">=</span> <span class="va">allele_counts_atac_ref</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="va">peaks</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rna_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  type <span class="op">=</span> <span class="st">"RNA"</span>,</span></span>
<span class="r-in"><span>  mat_counts <span class="op">=</span> <span class="va">mat_counts_rna_ref</span>,</span></span>
<span class="r-in"><span>  allele_counts <span class="op">=</span> <span class="va">allele_counts_rna_ref</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="va">genes</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create muscadet objects</span></span></span>
<span class="r-in"><span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscadetObject.html">CreateMuscadetObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  omics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">atac</span>, <span class="va">rna</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  bulk.lrr <span class="op">=</span> <span class="va">bulk_lrr</span>,</span></span>
<span class="r-in"><span>  bulk.label <span class="op">=</span> <span class="st">"WGS"</span>,</span></span>
<span class="r-in"><span>  genome <span class="op">=</span> <span class="st">"hg38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">muscadet_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscadetObject.html">CreateMuscadetObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  omics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">atac_ref</span>, <span class="va">rna_ref</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  genome <span class="op">=</span> <span class="st">"hg38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compute log R ratio for RNA</span></span></span>
<span class="r-in"><span><span class="va">obj_rna</span> <span class="op">&lt;-</span> <span class="fu">computeLogRatioRNA</span><span class="op">(</span></span></span>
<span class="r-in"><span>  matTumor <span class="op">=</span> <span class="fu"><a href="muscadet-methods.html">matCounts</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">$</span><span class="va">RNA</span>,</span></span>
<span class="r-in"><span>  matRef <span class="op">=</span> <span class="fu"><a href="muscadet-methods.html">matCounts</a></span><span class="op">(</span><span class="va">muscadet_ref</span><span class="op">)</span><span class="op">$</span><span class="va">RNA</span>,</span></span>
<span class="r-in"><span>  genesCoord <span class="op">=</span> <span class="fu"><a href="muscadet-methods.html">coordFeatures</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">$</span><span class="va">RNA</span>,</span></span>
<span class="r-in"><span>  genome <span class="op">=</span> <span class="fu">slot</span><span class="op">(</span><span class="va">muscadet</span>, <span class="st">"genome"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  refReads <span class="op">=</span> <span class="fl">2</span> <span class="co"># low value for example subsampled datasets</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 01 - Match genes in count matrix with coordinates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 02 - Filtering genes: Minimum of 2 read(s) in reference cells and minimum of 0.01 read(s) in average per reference cell</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 03 - Normalization for sequencing depth: Normalized counts per million</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 04 - Log transformation and normalization by reference data: log R ratio</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 05 - Capping the range of values: threshold = 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 06 - Smoothing values on gene windows: 101 genes per window</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 07 - Centering of cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 08 - Correcting by reference variability</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">obj_rna</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">keep</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FALSE  TRUE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   151   349 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># With results form every step when `all_steps = TRUE`</span></span></span>
<span class="r-in"><span><span class="va">obj_rna_all</span> <span class="op">&lt;-</span> <span class="fu">computeLogRatioRNA</span><span class="op">(</span></span></span>
<span class="r-in"><span>  matTumor <span class="op">=</span> <span class="fu"><a href="muscadet-methods.html">matCounts</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">$</span><span class="va">RNA</span>,</span></span>
<span class="r-in"><span>  matRef <span class="op">=</span> <span class="fu"><a href="muscadet-methods.html">matCounts</a></span><span class="op">(</span><span class="va">muscadet_ref</span><span class="op">)</span><span class="op">$</span><span class="va">RNA</span>,</span></span>
<span class="r-in"><span>  genesCoord <span class="op">=</span> <span class="fu"><a href="muscadet-methods.html">coordFeatures</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">$</span><span class="va">RNA</span>,</span></span>
<span class="r-in"><span>  genome <span class="op">=</span> <span class="fu">slot</span><span class="op">(</span><span class="va">muscadet</span>, <span class="st">"genome"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  refReads <span class="op">=</span> <span class="fl">2</span>, <span class="co"># low value for example subsampled datasets</span></span></span>
<span class="r-in"><span>  all_steps <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 01 - Match genes in count matrix with coordinates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 02 - Filtering genes: Minimum of 2 read(s) in reference cells and minimum of 0.01 read(s) in average per reference cell</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 03 - Normalization for sequencing depth: Normalized counts per million</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 04 - Log transformation and normalization by reference data: log R ratio</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 05 - Capping the range of values: threshold = 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 06 - Smoothing values on gene windows: 101 genes per window</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 07 - Centering of cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 08 - Correcting by reference variability</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">obj_rna_all</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "step01" "step02" "step03" "step04" "step05" "step06" "step07" "step08"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "params" "coord" </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">obj_rna_all</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">keep</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FALSE  TRUE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   151   349 </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">obj_rna_all</span><span class="op">$</span><span class="va">step08</span><span class="op">$</span><span class="va">matTumor</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 349</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marie Denoulet, Nils Giordano, Eric Letouzé, Céline Vallot, Stéphane Minvielle.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

