<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute log R ratios — computeLogRatio • muscadet</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute log R ratios — computeLogRatio"><meta name="description" content="Computes log R ratios from raw count matrices within
muscadet objects. The log R ratios values are computed
based on read counts from a sample muscadet object
versus read counts a the reference muscadet object.
In the output sample muscadet object, the newly computed
matrix of log R ratios is added for the selected omic."><meta property="og:description" content="Computes log R ratios from raw count matrices within
muscadet objects. The log R ratios values are computed
based on read counts from a sample muscadet object
versus read counts a the reference muscadet object.
In the output sample muscadet object, the newly computed
matrix of log R ratios is added for the selected omic."><meta property="og:image" content="https://icagen.github.io/muscadet/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">muscadet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/muscadet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/data-prep.html">Preparation of input data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ICAGEN/muscadet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Compute log R ratios</h1>
      <small class="dont-index">Source: <a href="https://github.com/ICAGEN/muscadet/blob/main/R/computeLogRatio.R" class="external-link"><code>R/computeLogRatio.R</code></a></small>
      <div class="d-none name"><code>computeLogRatio.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Computes log R ratios from raw count matrices within
<code><a href="muscadet-class.html">muscadet</a></code> objects. The log R ratios values are computed
based on read counts from a sample <code><a href="muscadet-class.html">muscadet</a></code> object
<em>versus</em> read counts a the <code>reference</code> <code><a href="muscadet-class.html">muscadet</a></code> object.
In the output sample <code><a href="muscadet-class.html">muscadet</a></code> object, the newly computed
matrix of log R ratios is added for the selected <code>omic</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">computeLogRatio</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">reference</span>,</span>
<span>  <span class="va">omic</span>,</span>
<span>  method <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  new.label.features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  remove.raw <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  all_steps <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A <code><a href="muscadet-class.html">muscadet</a></code> object containing sample data
(<code>muscadet</code>).</p></dd>


<dt id="arg-reference">reference<a class="anchor" aria-label="anchor" href="#arg-reference"></a></dt>
<dd><p>Another <code><a href="muscadet-class.html">muscadet</a></code> object containing
reference data (<code>muscadet</code>).</p></dd>


<dt id="arg-omic">omic<a class="anchor" aria-label="anchor" href="#arg-omic"></a></dt>
<dd><p>Name of the omic to apply this function (<code>character</code> string).</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>Method to apply to the selected omic (<code>character</code> string).
Supported methods are "ATAC" and "RNA":</p><ul><li><p>"ATAC" method calls for <code><a href="computeLogRatioATAC.html">computeLogRatioATAC()</a></code> function</p></li>
<li><p>"RNA" method calls for <code><a href="computeLogRatioRNA.html">computeLogRatioRNA()</a></code> function</p></li>
</ul><p>If <code>NULL</code> and if the omic type is either "ATAC" or "RNA", the corresponding
method will be applied.</p></dd>


<dt id="arg-new-label-features">new.label.features<a class="anchor" aria-label="anchor" href="#arg-new-label-features"></a></dt>
<dd><p>New label for features (<code>character</code> string). If
<code>NULL</code>, the label remains unchanged when using the "RNA" method and becomes
"windows of peaks" when using the "ATAC" method.</p></dd>


<dt id="arg-remove-raw">remove.raw<a class="anchor" aria-label="anchor" href="#arg-remove-raw"></a></dt>
<dd><p><code>TRUE</code> or <code>FALSE</code> (<code>logical</code>). Whether to remove raw count
matrices. <code>TRUE</code> by default to reduce object size. Setting it to <code>FALSE</code>
will keep raw count matrices within the object.</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, suppresses informative messages during
execution. Default is <code>FALSE</code>.</p></dd>


<dt id="arg-all-steps">all_steps<a class="anchor" aria-label="anchor" href="#arg-all-steps"></a></dt>
<dd><p>Logical value to indicate whether matrices at each step of
the log ratio computation are kept and returned (<code>logical</code>). IMPORTANT: if
<code>TRUE</code>, the function does not return an updated
<code><a href="muscadet-class.html">muscadet</a></code> object but a list with matrices at each step
of the log ratio computation (see Value section). Default is <code>FALSE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Arguments passed on to <code><a href="computeLogRatioATAC.html">computeLogRatioATAC</a></code>, <code><a href="computeLogRatioRNA.html">computeLogRatioRNA</a></code></p><dl><dt><code>windowSize</code></dt>
<dd><p>Size of windows in base pairs (<code>integer</code> value). By default: <code>10e6</code> (10
Mbp).</p></dd>

    <dt><code>slidingSize</code></dt>
<dd><p>Distance between start positions of sliding windows in base pairs
(<code>integer</code> value). If set to the same value as <code>windowSize</code>, the windows
don't overlap. By default: <code>2e6</code> (2 Mbp).</p></dd>

    <dt><code>minReads</code></dt>
<dd><p>Minimum read average per window in reference cells (<code>integer</code>
value). By default: <code>5</code>.</p></dd>

    <dt><code>minPeaks</code></dt>
<dd><p>Minimum number of peaks per window (<code>integer</code> value). By
default: <code>100</code>.</p></dd>

    <dt><code>genesPerWindow</code></dt>
<dd><p>Number of genes per moving window (<code>integer</code> value). By
default: <code>101</code>.</p></dd>

    <dt><code>refReads</code></dt>
<dd><p>Minimum of reads in reference cells (<code>integer</code> value). By
default: <code>100</code>.</p></dd>

    <dt><code>refMeanReads</code></dt>
<dd><p>Minimum of average reads per reference cell (<code>integer</code>
value). By default: <code>0.01</code>.</p></dd>

    <dt><code>thresh_capping</code></dt>
<dd><p>Threshold to cap the range of log R ratio values
(<code>numeric</code> value). By default: <code>3</code>.</p></dd>


</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code><a href="muscadet-class.html">muscadet</a></code> object corresponding to the sample
<code><a href="muscadet-class.html">muscadet</a></code> object (<code>x</code>) containing the computed log R ratio
matrix in the <code>coverage</code> slot of the selected <code>omic</code>.</p>
<p>If the <code>all_steps</code> argument is set to <code>TRUE</code>, it returns a list with intermediate
matrices <code>matTumor</code> and <code>matRef</code> from every step from <code>step01</code> to <code>step08</code>
(no <code>step06</code> for <code>method</code> = "ATAC"), <code>coord</code> table of features coordinates
and associated data, and <code>params</code> list of parameters used for the function.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Log R ratios computation steps are described in functions:</p><ul><li><p><code><a href="computeLogRatioATAC.html">computeLogRatioATAC()</a></code></p></li>
<li><p><code><a href="computeLogRatioRNA.html">computeLogRatioRNA()</a></code></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other computeLogRatio:
<code><a href="computeLogRatioATAC.html">computeLogRatioATAC</a>()</code>,
<code><a href="computeLogRatioRNA.html">computeLogRatioRNA</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Create muscomic objects</span></span></span>
<span class="r-in"><span><span class="va">atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  type <span class="op">=</span> <span class="st">"ATAC"</span>,</span></span>
<span class="r-in"><span>  mat_counts <span class="op">=</span> <span class="va">mat_counts_atac_tumor</span>,</span></span>
<span class="r-in"><span>  allele_counts <span class="op">=</span> <span class="va">allele_counts_atac_tumor</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="va">peaks</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  type <span class="op">=</span> <span class="st">"RNA"</span>,</span></span>
<span class="r-in"><span>  mat_counts <span class="op">=</span> <span class="va">mat_counts_rna_tumor</span>,</span></span>
<span class="r-in"><span>  allele_counts <span class="op">=</span> <span class="va">allele_counts_rna_tumor</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="va">genes</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">atac_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  type <span class="op">=</span> <span class="st">"ATAC"</span>,</span></span>
<span class="r-in"><span>  mat_counts <span class="op">=</span> <span class="va">mat_counts_atac_ref</span>,</span></span>
<span class="r-in"><span>  allele_counts <span class="op">=</span> <span class="va">allele_counts_atac_ref</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="va">peaks</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rna_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  type <span class="op">=</span> <span class="st">"RNA"</span>,</span></span>
<span class="r-in"><span>  mat_counts <span class="op">=</span> <span class="va">mat_counts_rna_ref</span>,</span></span>
<span class="r-in"><span>  allele_counts <span class="op">=</span> <span class="va">allele_counts_rna_ref</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="va">genes</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create muscadet objects</span></span></span>
<span class="r-in"><span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscadetObject.html">CreateMuscadetObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  omics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">atac</span>, <span class="va">rna</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  bulk.lrr <span class="op">=</span> <span class="va">bulk_lrr</span>,</span></span>
<span class="r-in"><span>  bulk.label <span class="op">=</span> <span class="st">"WGS"</span>,</span></span>
<span class="r-in"><span>  genome <span class="op">=</span> <span class="st">"hg38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">muscadet_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="CreateMuscadetObject.html">CreateMuscadetObject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  omics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">atac_ref</span>, <span class="va">rna_ref</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  genome <span class="op">=</span> <span class="st">"hg38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># compute log R ratios for ATAC</span></span></span>
<span class="r-in"><span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu">computeLogRatio</span><span class="op">(</span></span></span>
<span class="r-in"><span>  x <span class="op">=</span> <span class="va">muscadet</span>,</span></span>
<span class="r-in"><span>  reference <span class="op">=</span> <span class="va">muscadet_ref</span>,</span></span>
<span class="r-in"><span>  omic <span class="op">=</span> <span class="st">"ATAC"</span>,</span></span>
<span class="r-in"><span>  method <span class="op">=</span> <span class="st">"ATAC"</span>,</span></span>
<span class="r-in"><span>  minReads <span class="op">=</span> <span class="fl">1</span>, <span class="co"># low value for example subsampled datasets</span></span></span>
<span class="r-in"><span>  minPeaks <span class="op">=</span> <span class="fl">1</span> <span class="co"># low value for example subsampled datasets</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -- computeLogRatio: Method 'ATAC' using computeLogRatioATAC().</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 01 - Group peaks in windows: window size set at 10 Mb, sliding by 2 Mb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 02 - Filtering windows: Minimum of 1 peaks per window with a minimum average of 1 read(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 03 - Normalization for sequencing depth: Normalized counts per million</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 04 - Log transformation and normalization by reference data: log R ratio</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 05 - Capping the range of values: threshold = 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 06 - [No step 06 for scATAC-seq]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 07 - Centering of cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 08 - Correcting by reference variability</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># compute log R ratios for RNA</span></span></span>
<span class="r-in"><span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu">computeLogRatio</span><span class="op">(</span></span></span>
<span class="r-in"><span>  x <span class="op">=</span> <span class="va">muscadet</span>,</span></span>
<span class="r-in"><span>  reference <span class="op">=</span> <span class="va">muscadet_ref</span>,</span></span>
<span class="r-in"><span>  omic <span class="op">=</span> <span class="st">"RNA"</span>,</span></span>
<span class="r-in"><span>  method <span class="op">=</span> <span class="st">"RNA"</span>,</span></span>
<span class="r-in"><span>  refReads <span class="op">=</span> <span class="fl">2</span> <span class="co"># low value for example subsampled datasets</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -- computeLogRatio: Method 'RNA' using computeLogRatioRNA().</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 01 - Match genes in count matrix with coordinates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 02 - Filtering genes: Minimum of 2 read(s) in reference cells and minimum of 0.01 read(s) in average per reference cell</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 03 - Normalization for sequencing depth: Normalized counts per million</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 04 - Log transformation and normalization by reference data: log R ratio</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 05 - Capping the range of values: threshold = 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 06 - Smoothing values on gene windows: 101 genes per window</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 07 - Centering of cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 08 - Correcting by reference variability</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done.</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marie Denoulet, Nils Giordano, Eric Letouzé, Céline Vallot, Stéphane Minvielle.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

