<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Process read count matrix and segmentation — preProcSample2 • muscadet</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Process read count matrix and segmentation — preProcSample2"><meta name="description" content="Function adapted from facets::preProcSample() to process read count matrix
and generates a segmentation tree. The modifications from the original
function includes:
Incorporation of cluster and signal columns into the final result.
Change in the log ratio (cnlr) for allelic data, it is computed as the
log ratio (cnlr) mean between the previous and next coverage positions.

"><meta property="og:description" content="Function adapted from facets::preProcSample() to process read count matrix
and generates a segmentation tree. The modifications from the original
function includes:
Incorporation of cluster and signal columns into the final result.
Change in the log ratio (cnlr) for allelic data, it is computed as the
log ratio (cnlr) mean between the previous and next coverage positions.

"><meta property="og:image" content="https://icagen.github.io/muscadet/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">muscadet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/muscadet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ICAGEN/muscadet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Process read count matrix and segmentation</h1>
      <small class="dont-index">Source: <a href="https://github.com/ICAGEN/muscadet/blob/main/R/cnaCalling.R" class="external-link"><code>R/cnaCalling.R</code></a></small>
      <div class="d-none name"><code>preProcSample2.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Function adapted from <code><a href="https://rdrr.io/pkg/facets/man/preProcSample.html" class="external-link">facets::preProcSample()</a></code> to process read count matrix
and generates a segmentation tree. The modifications from the original
function includes:</p><ul><li><p>Incorporation of <code>cluster</code> and <code>signal</code> columns into the final result.</p></li>
<li><p>Change in the log ratio (<code>cnlr</code>) for allelic data, it is computed as the
log ratio (<code>cnlr</code>) mean between the previous and next coverage positions.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">preProcSample2</span><span class="op">(</span></span>
<span>  <span class="va">rcmat</span>,</span>
<span>  het.thresh <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  snp.nbhd <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  cval <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  gbuild <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  hetscale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ndepth <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  ndepthmax <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p>This function is derived from the <code><a href="https://rdrr.io/pkg/facets/man/preProcSample.html" class="external-link">facets::preProcSample()</a></code> function,
with the modifications to fit its use in muscadet.</p>
<p>Seshan VE, Shen R (2021). <em>facets: Cellular Fraction and Copy Numbers from Tumor Sequencing</em>.
R package version 0.6.2, <a href="https://github.com/mskcc/facets" class="external-link">https://github.com/mskcc/facets</a>.</p>
    </div>
    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-rcmat">rcmat<a class="anchor" aria-label="anchor" href="#arg-rcmat"></a></dt>
<dd><p>A data frame with 8 required columns: <code>Chrom</code>, <code>Pos</code>, <code>NOR.DP</code>,
<code>NOR.RD</code>, <code>TUM.DP</code>, <code>TUM.RD</code>, <code>cluster</code>, and <code>signal</code> (<code>data.frame</code>).</p></dd>


<dt id="arg-het-thresh">het.thresh<a class="anchor" aria-label="anchor" href="#arg-het-thresh"></a></dt>
<dd><p>VAF (Variant Allele Frequency) threshold to call variant positions
heterozygous (<code>numeric</code>). Default: 0.25.</p></dd>


<dt id="arg-snp-nbhd">snp.nbhd<a class="anchor" aria-label="anchor" href="#arg-snp-nbhd"></a></dt>
<dd><p>Window size for selecting loci to reduce serial correlation
(<code>numeric</code>). Default: 250.</p></dd>


<dt id="arg-cval">cval<a class="anchor" aria-label="anchor" href="#arg-cval"></a></dt>
<dd><p>Critical value for segmentation (<code>numeric</code>). Default: 25.</p></dd>


<dt id="arg-gbuild">gbuild<a class="anchor" aria-label="anchor" href="#arg-gbuild"></a></dt>
<dd><p>Genome build used for alignment. One of <code>"hg19"</code>,
<code>"hg38"</code>, or <code>"mm10"</code> (<code>character</code>). Default: <code>"hg19"</code>.</p></dd>


<dt id="arg-hetscale">hetscale<a class="anchor" aria-label="anchor" href="#arg-hetscale"></a></dt>
<dd><p>Logical value indicating whether log odds ratio (logOR)
should be scaled to give more weight in the test statistics for
segmentation and clustering (<code>logical</code>). Usually only 10% of snps are hets
and hetscale gives the logOR contribution to T-square as 0.25/proportion of
hets. Default: <code>TRUE</code>.</p></dd>


<dt id="arg-ndepth">ndepth<a class="anchor" aria-label="anchor" href="#arg-ndepth"></a></dt>
<dd><p>Minimum depth in normal reference to keep (<code>numeric</code>). Default:
5.</p></dd>


<dt id="arg-ndepthmax">ndepthmax<a class="anchor" aria-label="anchor" href="#arg-ndepthmax"></a></dt>
<dd><p>Maximum normal coverage threshold for filtering loci
(<code>numeric</code>). Default: 1000.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing:</p><dl><dt>pmat</dt>
<dd><p>Read counts and other elements of all the loci.</p></dd>

<dt>gbuild</dt>
<dd><p>Genome build used for the analysis.</p></dd>

<dt>nX</dt>
<dd><p>Chromosome number for X (e.g., 23 for human, 20 for mouse).</p></dd>

<dt>clusters</dt>
<dd><p>Unique clusters from the processed data.</p></dd>

<dt>seg.tree</dt>
<dd><p>Segmentation tree for each chromosome.</p></dd>

<dt>jointseg</dt>
<dd><p>Segmented variant positions data.</p></dd>

<dt>hscl</dt>
<dd><p>Scaling factor for logOR data.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function processes variant positions data to generate a segmentation
tree. It uses <code><a href="https://rdrr.io/pkg/facets/man/facets-internal.html" class="external-link">procSnps</a></code> to compute initial values,
adjusts the log ratio for allelic signals, and computes segmentation using
<code><a href="https://rdrr.io/pkg/facets/man/facets-internal.html" class="external-link">segsnps</a></code>.</p>
<p>SNPs (or variants) in a genome are not evenly spaced, and loci are sampled
within the specified window (<code>snp.nbhd</code>) to reduce serial correlation.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<dl><dt><a href="https://rdrr.io/pkg/facets/man/facets-package.html" class="external-link">facets-package</a> package</dt>
<dd><p>Shen R, Seshan VE. FACETS: allele-specific copy number and
clonal heterogeneity analysis tool for high-throughput DNA sequencing.
Nucleic Acids Res. 2016 Sep 19;44(16):e131.
doi: <a href="https://www.doi.org/10.1093/nar/gkw520" class="external-link">10.1093/nar/gkw520</a>.
PMID: 27270079; PMCID: PMC5027494.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/facets/man/preProcSample.html" class="external-link">preProcSample</a></code>, <code><a href="https://rdrr.io/pkg/facets/man/facets-internal.html" class="external-link">procSnps</a></code>,
<code><a href="https://rdrr.io/pkg/facets/man/facets-internal.html" class="external-link">segsnps</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"facets"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Load example muscadet object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">muscadet_obj</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">muscadet_obj</span><span class="op">$</span><span class="va">cnacalling</span><span class="op">$</span><span class="va">combined.counts</span></span></span>
<span class="r-in"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html" class="external-link">complete.cases</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">counts_clus</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">counts</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">preProcSample2</span><span class="op">(</span><span class="va">counts_clus</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marie Denoulet, Nils Giordano, Eric Letouzé, Céline Vallot, Stéphane Minvielle.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

