<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Preparation of input data • muscadet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="data-prep_files/libs/quarto-html/popper.min.js"></script><script src="data-prep_files/libs/quarto-html/tippy.umd.min.js"></script><link href="data-prep_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="data-prep_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Preparation of input data">
<meta property="og:image" content="https://icagen.github.io/muscadet/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">muscadet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/muscadet.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/data-prep.html">Preparation of input data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ICAGEN/muscadet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Preparation of input data</h1>




      <small class="dont-index">Source: <a href="https://github.com/ICAGEN/muscadet/blob/main/vignettes/data-prep.qmd" class="external-link"><code>vignettes/data-prep.qmd</code></a></small>
      <div class="d-none name"><code></code></div>
    </div>






    <section class="section level2" data-number="1"><h2 data-number="1" id="sec-mat-counts">
<span class="header-section-number">1</span> Count matrices<a class="anchor" aria-label="anchor" href="#sec-mat-counts"></a>
</h2>
<p>The <code>mat_counts</code> input is required for muscadet analysis and must be provided with <code><a href="../reference/CreateMuscomicObject.html">CreateMuscomicObject()</a></code>.</p>
<p>This input should be a matrix (sparse <code>dgCMatrix</code> or not) of raw read counts with features as rows and cells as columns (<a href="#tbl-mat-counts" class="quarto-xref">Table 1</a>).</p>
<p>See <code><a href="../reference/mat_counts.html">?mat_counts</a></code> for full documentation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"mat_counts_atac_tumor"</span><span class="op">)</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="va">mat_counts_atac_tumor</span><span class="op">[</span><span class="fl">698</span><span class="op">:</span><span class="fl">700</span>,<span class="fl">34</span><span class="op">:</span><span class="fl">36</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div id="tbl-mat-counts" class="cell quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-tbl"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-mat-counts-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table 1: Allele counts table example
</figcaption><div aria-describedby="tbl-mat-counts-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="table do-not-create-environment cell caption-top">
<colgroup>
<col style="width: 20%">
<col style="width: 26%">
<col style="width: 26%">
<col style="width: 26%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">samplename_CCGTTGCGTGTGTCCC-1</th>
<th style="text-align: right;">samplename_CCTGGATCATTCCTCG-1</th>
<th style="text-align: right;">samplename_CCTTCAATCCGCCAAA-1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">12_118016401_118016901</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">12_124564171_124564671</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">12_132710447_132710947</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div>
</div>
<ul>
<li>From Seurat/Signac analysis:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span> <span class="co"># Seurat v5</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"atac_small"</span><span class="op">)</span> <span class="co"># Example Seurat class object from Signac</span></span>
<span></span>
<span><span class="co"># dgCMatrix matrix from Assay named "RNA"</span></span>
<span><span class="va">mat_counts_RNA</span> <span class="op">&lt;-</span> <span class="va">atac_small</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span> </span>
<span><span class="co"># dgCMatrix matrix from ChromatinAssay named "peaks"</span></span>
<span><span class="va">mat_counts_ATAC</span> <span class="op">&lt;-</span> <span class="va">atac_small</span><span class="op">[[</span><span class="st">"peaks"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span> </span></code></pre></div>
</div>
<ul>
<li>From ArchR/SummarizedExperiment analysis:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span><span class="co"># RangedSummarizedExperiment class object</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"&lt;SummarizedExperiment_path&gt;"</span><span class="op">)</span> </span>
<span><span class="co"># dgCMatrix matrix</span></span>
<span><span class="va">mat_counts_ATAC</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> </span></code></pre></div>
</div>
</section><section class="section level2" data-number="2"><h2 data-number="2" id="sec-allele-counts">
<span class="header-section-number">2</span> Allele counts tables<a class="anchor" aria-label="anchor" href="#sec-allele-counts"></a>
</h2>
<p>The <code>allele_counts</code> input is optional for the clustering step (allelic information not used for clustering) but required for the CNA calling step: it can be provided in the initial object creation step with <code><a href="../reference/CreateMuscomicObject.html">CreateMuscomicObject()</a></code> or added to <code>muscadet</code> objects later on with <code><a href="../reference/addAlleleCounts.html">addAlleleCounts()</a></code>.</p>
<p>This input should be in a data frame format containing allelic read counts at specific variant positions across cells (<a href="#tbl-allele-counts" class="quarto-xref">Table 2</a>). The format follows a Variant Calling Format (VCF)-like structure and should include:</p>
<ul>
<li>
<code>cell</code>: unique cell barcode</li>
<li>
<code>id</code>: variant identifier (e.g. CHROM_POS_REF_ALT)</li>
<li>
<code>CHROM</code>: chromosome</li>
<li>
<code>POS</code>: position</li>
<li>
<code>REF</code> / <code>ALT</code>: reference and alternate alleles</li>
<li>
<code>RD</code> / <code>AD</code>: counts for reference and alternate alleles</li>
<li>
<code>DP</code>: total depth</li>
<li>
<code>GT</code>: genotype <em>(optional)</em>
</li>
</ul>
<p>See <code><a href="../reference/allele_counts.html">?allele_counts</a></code> for full documentation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"allele_counts_atac_tumor"</span><span class="op">)</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">allele_counts_atac_tumor</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div id="tbl-allele-counts" class="cell quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-tbl"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-allele-counts-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table 2: Allele counts table example
</figcaption><div aria-describedby="tbl-allele-counts-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="table do-not-create-environment cell caption-top">
<colgroup>
<col style="width: 6%">
<col style="width: 32%">
<col style="width: 16%">
<col style="width: 6%">
<col style="width: 9%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 9%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">cell</th>
<th style="text-align: left;">id</th>
<th style="text-align: right;">CHROM</th>
<th style="text-align: right;">POS</th>
<th style="text-align: left;">REF</th>
<th style="text-align: left;">ALT</th>
<th style="text-align: right;">RD</th>
<th style="text-align: right;">AD</th>
<th style="text-align: right;">DP</th>
<th style="text-align: left;">GT</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">4694</td>
<td style="text-align: left;">samplename_GAAAGGCTCTAAGTGC-1</td>
<td style="text-align: left;">1_2323263_T_G</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2323263</td>
<td style="text-align: left;">T</td>
<td style="text-align: left;">G</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">1|0</td>
</tr>
<tr class="even">
<td style="text-align: left;">13631</td>
<td style="text-align: left;">samplename_ACTTACAAGTAAGGGC-1</td>
<td style="text-align: left;">1_4007524_T_C</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">4007524</td>
<td style="text-align: left;">T</td>
<td style="text-align: left;">C</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">1|0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">24816</td>
<td style="text-align: left;">samplename_GTAGCGCTCCTTTACG-1</td>
<td style="text-align: left;">1_6599445_C_A</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6599445</td>
<td style="text-align: left;">C</td>
<td style="text-align: left;">A</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">1|0</td>
</tr>
<tr class="even">
<td style="text-align: left;">31038</td>
<td style="text-align: left;">samplename_GTACTTCGTTGTGACA-1</td>
<td style="text-align: left;">1_8201062_C_T</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">8201062</td>
<td style="text-align: left;">C</td>
<td style="text-align: left;">T</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">0|1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">34534</td>
<td style="text-align: left;">samplename_AAGCTCCCAACCCTCC-1</td>
<td style="text-align: left;">1_9491207_C_A</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">9491207</td>
<td style="text-align: left;">C</td>
<td style="text-align: left;">A</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">0|1</td>
</tr>
<tr class="even">
<td style="text-align: left;">49929</td>
<td style="text-align: left;">samplename_ATGGTTATCCAAATCA-1</td>
<td style="text-align: left;">1_15417791_C_A</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">15417791</td>
<td style="text-align: left;">C</td>
<td style="text-align: left;">A</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">0|1</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div>
</div>
<section class="level2" data-number="2.1"><h3 data-number="2.1" id="variant-positions">
<span class="header-section-number">2.1</span> Variant positions<a class="anchor" aria-label="anchor" href="#variant-positions"></a>
</h3>
<p>Variant positions can be derived from:</p>
<ul>
<li>Matched bulk sequencing to identify individual-specific heterozygous SNPs, or</li>
<li>Reference panels of common SNPs (e.g. gnomAD, 1000G).</li>
</ul>
<section class="level3" data-number="2.1.1"><h4 data-number="2.1.1" id="individual-specific-heterozygous-positions-from-bulk-data">
<span class="header-section-number">2.1.1</span> Individual-specific heterozygous positions from bulk data<a class="anchor" aria-label="anchor" href="#individual-specific-heterozygous-positions-from-bulk-data"></a>
</h4>
<p>Positions can be retrieved by running <a href="https://github.com/mskcc/facets" class="external-link">FACETS</a><a class="footnote-ref" role="doc-noteref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Shen R, Seshan VE. &lt;strong&gt;FACETS: allele-specific copy number and clonal heterogeneity analysis tool for high-throughput DNA sequencing. &lt;/strong&gt; &lt;em&gt;Nucleic Acids Res&lt;/em&gt; (2016). &lt;a href="https://www.doi.org/10.1093/nar/gkw520" class="external-link"&gt;https://www.doi.org/10.1093/nar/gkw520&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> on matched WGS/WES normal samples.</p>
<p>Use <code>snp-pileup</code> to extract allele counts from bulk BAM files at known sites (VCF of common SNPs from the <a href="https://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606/VCF/00-common_all.vcf.gz" class="external-link">NCBI database</a>). The output of <code>snp-pileup</code> can be filtered on depth and allele frequency to keep only heterozygous positions well covered.</p>
<p>See <a href="https://github.com/mskcc/facets/blob/master/inst/extcode/README.txt" class="external-link">FACETS snp-pileup documentation</a>.</p>
</section><section class="level3" data-number="2.1.2"><h4 data-number="2.1.2" id="panels-of-common-snps">
<span class="header-section-number">2.1.2</span> Panels of common SNPs<a class="anchor" aria-label="anchor" href="#panels-of-common-snps"></a>
</h4>
<ul>
<li>
<a href="https://gnomad.broadinstitute.org/" class="external-link">gnomAD</a> database (4,099 SNPs): <a href="https://cloud.google.com/life-sciences/docs/resources/public-datasets/gnomad?hl=en" class="external-link">data here</a>
</li>
<li>
<a href="https://www.internationalgenome.org/" class="external-link">1000G</a> database (2,548 SNPs): <a href="https://sourceforge.net/projects/cellsnp/files/SNPlist" class="external-link">data here</a>
</li>
</ul></section></section><section class="level2" data-number="2.2"><h3 data-number="2.2" id="count-reads-from-single-cells">
<span class="header-section-number">2.2</span> Count reads from single cells<a class="anchor" aria-label="anchor" href="#count-reads-from-single-cells"></a>
</h3>
<p><a href="https://horvathlab.github.io/NGS/SCReadCounts/" class="external-link">SCReadCounts</a><a class="footnote-ref" role="doc-noteref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Prashant, N.M., Alomran, N., Chen, Y. et al. &lt;strong&gt;SCReadCounts: estimation of cell-level SNVs expression from scRNA-seq data.&lt;/strong&gt; &lt;em&gt;BMC Genomics&lt;/em&gt; (2021). &lt;a href="https://doi.org/10.1186/s12864-021-07974-8" class="external-link"&gt;https://doi.org/10.1186/s12864-021-07974-8&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a> is used to get read counts per allele from single cell BAM files.</p>
<blockquote>
<p>The program <code>scReadCounts</code> manages the sequential execution of programs <code>readCounts</code> and <code>readCountsMatrix</code>, collects the necessary arguments for successful execution, and avoids unnecessary execution of the expensive <code>readCounts</code> tool if possible. <code>readCounts</code> requires three input files: a pooled single cell alignment, a list of genomic positions of interest, and the barcodes file (barcodes.tsv) . Optionally, <code>readCounts</code> can be user-configured for <em>read filtering</em> and <em>cell-barcode handling</em>, including restriction to barcodes of interest (achieved through specifying barcodes file - barcodes.tsv). <code>readCounts</code> utilizes the barcode information from the pooled single cell alignments and <strong>outputs the variant and reference read counts (nvar and nref, respectively), for each barcode (cell), restricted to those present in the barcodes file, in a tab separated text file</strong>. This file is then used as an input for the second program - <code>readCountsMatrix</code> - which, upon providing an output prefix, generates two outputs: (1) a cell-position matrix with absolute nvar and nref counts, and (2) a cell-position matrix with the expressed VAFRNA. VAFRNA is estimated at a user-defined threshold of minimum required sequencing reads (minR); default minR = 5. <code>readCountsMatrix</code> is time-efficient and can be re-run multiple times at various minR thresholds. Together, these tools facilitate single-cell level assessment of read counts. <em>From <a href="https://horvathlab.github.io/NGS/SCReadCounts/" class="external-link">https://horvathlab.github.io/NGS/SCReadCounts/</a></em></p>
</blockquote>
<p>As the desired output is the tab separated text file with variant and reference read counts for each barcode, the command <a href="https://horvathlab.github.io/NGS/ReadCounts/" class="external-link"><code>readCounts</code></a> is preferred to avoid unnecessary large matrix outputs from <code>readCountsMatrix</code>.</p>
<ul>
<li>
<a href="https://horvathlab.github.io/NGS/ReadCounts/docs/Usage.html" class="external-link">Command usage</a> for<code>readCounts</code>
</li>
<li>Details on <a href="https://horvathlab.github.io/NGS/ReadCounts/docs/Grouping.html" class="external-link">Read grouping</a>
</li>
<li>Details on <a href="https://horvathlab.github.io/NGS/ReadCounts/docs/Filtering.html" class="external-link">Alignment filter</a>
</li>
<li>Example of <code>readCounts</code> command, with <code>STARsolo_CB</code> read grouping and <code>MPileup</code> alignment filter:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">readCounts</span> <span class="dt">\</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">-s</span> <span class="op">&lt;</span>vcf_file.vcf<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-r</span> <span class="op">&lt;</span>bam_file.bam<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-b</span> None <span class="dt">\</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">-G</span> STARsolo_CB <span class="dt">\</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">-f</span> MPileup <span class="dt">\</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> <span class="op">&lt;</span>output_file.tsv<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">-t</span> <span class="op">&lt;</span>threads<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>    <span class="op">&amp;&gt;</span> <span class="op">&lt;</span>log_file<span class="op">&gt;</span></span></code></pre></div>
</div>
<p>Then, the output of <code>readCounts</code> (table of variant and reference read counts) can be formatted to fit <code>muscadet</code> input requirement with <code><a href="../reference/process_allele.html">process_allele()</a></code> (<a href="#tbl-process-allele1" class="quarto-xref">Table 3</a>, <a href="#tbl-process-allele2" class="quarto-xref">Table 4</a>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example data frame of readCounts results</span></span>
<span><span class="va">readcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  CHROM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"1"</span>, <span class="st">"2"</span><span class="op">)</span>,</span>
<span>  POS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10101</span>, <span class="fl">20202</span>, <span class="fl">30303</span><span class="op">)</span>,</span>
<span>  REF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"G"</span>, <span class="st">"T"</span><span class="op">)</span>,</span>
<span>  ALT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G"</span>, <span class="st">"A"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>  ReadGroup <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell1"</span>, <span class="st">"cell1"</span>, <span class="st">"cell1"</span><span class="op">)</span>,</span>
<span>  SNVCountForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  SNVCountReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">6</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  RefCountForward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">15</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  RefCountReverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">18</span>, <span class="fl">12</span>, <span class="fl">8</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">readcounts</span><span class="op">$</span><span class="va">SNVCount</span> <span class="op">&lt;-</span> <span class="va">readcounts</span><span class="op">$</span><span class="va">SNVCountForward</span> <span class="op">+</span> <span class="va">readcounts</span><span class="op">$</span><span class="va">SNVCountReverse</span></span>
<span><span class="va">readcounts</span><span class="op">$</span><span class="va">RefCount</span> <span class="op">&lt;-</span> <span class="va">readcounts</span><span class="op">$</span><span class="va">RefCountForward</span> <span class="op">+</span><span class="va">readcounts</span><span class="op">$</span><span class="va">RefCountReverse</span></span>
<span><span class="va">readcounts</span><span class="op">$</span><span class="va">GoodReads</span> <span class="op">&lt;-</span> <span class="va">readcounts</span><span class="op">$</span><span class="va">SNVCount</span> <span class="op">+</span> <span class="va">readcounts</span><span class="op">$</span><span class="va">RefCount</span></span>
<span><span class="va">readcounts</span><span class="op">[</span>, <span class="st">"%BadRead"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">readcounts</span><span class="op">$</span><span class="va">VAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">readcounts</span><span class="op">$</span><span class="va">SNVCount</span> <span class="op">/</span> <span class="va">readcounts</span><span class="op">$</span><span class="va">GoodReads</span>, <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">kable</span><span class="op">(</span><span class="va">readcounts</span><span class="op">)</span></span></code></pre></div>
<div id="tbl-process-allele1" class="cell quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-tbl"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-process-allele1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table 3: Example of readCounts output
</figcaption><div aria-describedby="tbl-process-allele1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="table do-not-create-environment cell caption-top" style="width:100%;">
<colgroup>
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 7%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 5%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">CHROM</th>
<th style="text-align: right;">POS</th>
<th style="text-align: left;">REF</th>
<th style="text-align: left;">ALT</th>
<th style="text-align: left;">ReadGroup</th>
<th style="text-align: right;">SNVCountForward</th>
<th style="text-align: right;">SNVCountReverse</th>
<th style="text-align: right;">RefCountForward</th>
<th style="text-align: right;">RefCountReverse</th>
<th style="text-align: right;">SNVCount</th>
<th style="text-align: right;">RefCount</th>
<th style="text-align: right;">GoodReads</th>
<th style="text-align: right;">%BadRead</th>
<th style="text-align: right;">VAF</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1</td>
<td style="text-align: right;">10101</td>
<td style="text-align: left;">A</td>
<td style="text-align: left;">G</td>
<td style="text-align: left;">cell1</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">20</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">38</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.1915</td>
</tr>
<tr class="even">
<td style="text-align: left;">1</td>
<td style="text-align: right;">20202</td>
<td style="text-align: left;">G</td>
<td style="text-align: left;">A</td>
<td style="text-align: left;">cell1</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">43</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.3721</td>
</tr>
<tr class="odd">
<td style="text-align: left;">2</td>
<td style="text-align: right;">30303</td>
<td style="text-align: left;">T</td>
<td style="text-align: left;">C</td>
<td style="text-align: left;">cell1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">23</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.2174</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="../reference/process_allele.html">process_allele</a></span><span class="op">(</span><span class="va">readcounts</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div id="tbl-process-allele2" class="cell quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-tbl"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-process-allele2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table 4: Formatted table of allele counts
</figcaption><div aria-describedby="tbl-process-allele2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="table do-not-create-environment cell caption-top">
<thead><tr class="header">
<th style="text-align: left;">cell</th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">CHROM</th>
<th style="text-align: right;">POS</th>
<th style="text-align: left;">REF</th>
<th style="text-align: left;">ALT</th>
<th style="text-align: right;">RD</th>
<th style="text-align: right;">AD</th>
<th style="text-align: right;">DP</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">cell1</td>
<td style="text-align: left;">1_10101_A_G</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">10101</td>
<td style="text-align: left;">A</td>
<td style="text-align: left;">G</td>
<td style="text-align: right;">38</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">47</td>
</tr>
<tr class="even">
<td style="text-align: left;">cell1</td>
<td style="text-align: left;">1_20202_G_A</td>
<td style="text-align: left;">1</td>
<td style="text-align: right;">20202</td>
<td style="text-align: left;">G</td>
<td style="text-align: left;">A</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">43</td>
</tr>
<tr class="odd">
<td style="text-align: left;">cell1</td>
<td style="text-align: left;">2_30303_T_C</td>
<td style="text-align: left;">2</td>
<td style="text-align: right;">30303</td>
<td style="text-align: left;">T</td>
<td style="text-align: left;">C</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">23</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div>
</div>
</section></section><section class="section level2" data-number="3"><h2 data-number="3" id="sec-features">
<span class="header-section-number">3</span> Feature coordinates<a class="anchor" aria-label="anchor" href="#sec-features"></a>
</h2>
<p>A table of features coordinates with 4 columns (<code>CHROM</code>, <code>start</code>, <code>end</code>, <code>id</code>) matching features of count matrices (section <a href="#sec-mat-counts" class="quarto-xref">Section 1</a>) is required (<a href="#tbl-features" class="quarto-xref">Table 5</a>).</p>
<p>See <code><a href="../reference/features.html">?features</a></code> for full documentation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"peaks"</span><span class="op">)</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div id="tbl-features" class="cell quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-tbl"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-features-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table 5: Features coordinates table example
</figcaption><div aria-describedby="tbl-features-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="table do-not-create-environment cell caption-top">
<thead><tr class="header">
<th style="text-align: left;">CHROM</th>
<th style="text-align: right;">start</th>
<th style="text-align: right;">end</th>
<th style="text-align: left;">id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1</td>
<td style="text-align: right;">941542</td>
<td style="text-align: right;">942042</td>
<td style="text-align: left;">1_941542_942042</td>
</tr>
<tr class="even">
<td style="text-align: left;">1</td>
<td style="text-align: right;">3976396</td>
<td style="text-align: right;">3976896</td>
<td style="text-align: left;">1_3976396_3976896</td>
</tr>
<tr class="odd">
<td style="text-align: left;">1</td>
<td style="text-align: right;">6465944</td>
<td style="text-align: right;">6466444</td>
<td style="text-align: left;">1_6465944_6466444</td>
</tr>
<tr class="even">
<td style="text-align: left;">1</td>
<td style="text-align: right;">6901466</td>
<td style="text-align: right;">6901966</td>
<td style="text-align: left;">1_6901466_6901966</td>
</tr>
<tr class="odd">
<td style="text-align: left;">1</td>
<td style="text-align: right;">7402483</td>
<td style="text-align: right;">7402983</td>
<td style="text-align: left;">1_7402483_7402983</td>
</tr>
<tr class="even">
<td style="text-align: left;">1</td>
<td style="text-align: right;">10794813</td>
<td style="text-align: right;">10795313</td>
<td style="text-align: left;">1_10794813_10795313</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div>
</div>
<section class="level2" data-number="3.1"><h3 data-number="3.1" id="peaks">
<span class="header-section-number">3.1</span> Peaks<a class="anchor" aria-label="anchor" href="#peaks"></a>
</h3>
<ul>
<li>From Seurat/Signac</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="co"># Example Seurat class object from Signac</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"atac_small"</span><span class="op">)</span> </span>
<span><span class="co"># GRanges class object</span></span>
<span><span class="va">peaks_coord</span> <span class="op">&lt;-</span> <span class="va">atac_small</span><span class="op">[[</span><span class="st">"peaks"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ranges</span> </span>
<span></span>
<span><span class="va">peaks_coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">peaks_coord</span><span class="op">)</span> </span>
<span><span class="va">peaks_coord</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">peaks_coord</span><span class="op">$</span><span class="va">seqnames</span>,</span>
<span>                        <span class="va">peaks_coord</span><span class="op">$</span><span class="va">start</span>,</span>
<span>                        <span class="va">peaks_coord</span><span class="op">$</span><span class="va">end</span>,</span>
<span>                        sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="va">peaks_coord</span> <span class="op">&lt;-</span> <span class="va">peaks_coord</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqnames"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"id"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">peaks_coord</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHROM"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"id"</span><span class="op">)</span></span></code></pre></div>
</div>
<ul>
<li>From ArchR/SummarizedExperiment:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span><span class="co"># RangedSummarizedExperiment class object</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"&lt;SummarizedExperiment_path&gt;"</span><span class="op">)</span> </span>
<span><span class="co"># GRanges class object</span></span>
<span><span class="va">peaks_coord</span> <span class="op">&lt;-</span> <span class="fu">rowRanges</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">peaks_coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">peaks_coord</span><span class="op">)</span></span>
<span><span class="va">peaks_coord</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">peaks_coord</span><span class="op">$</span><span class="va">seqnames</span>, <span class="va">peaks_coord</span><span class="op">$</span><span class="va">start</span>, <span class="va">peaks_coord</span><span class="op">$</span><span class="va">end</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="va">peaks_coord</span> <span class="op">&lt;-</span> <span class="va">peaks_coord</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqnames"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"id"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">peaks_coord</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHROM"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"id"</span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2" data-number="3.2"><h3 data-number="3.2" id="genes">
<span class="header-section-number">3.2</span> Genes<a class="anchor" aria-label="anchor" href="#genes"></a>
</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use your own annotation</span></span>
<span><span class="va">genes_coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span> <span class="st">"&lt;genes_gtf_file&gt;"</span><span class="op">)</span></span>
<span><span class="va">genes_coord</span> <span class="op">&lt;-</span> <span class="va">genes_coord</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqnames"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"gene_name"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">genes_coord</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHROM"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or get coordinates </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span> <span class="co"># for human</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationDbi</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat_counts_RNA</span><span class="op">)</span></span>
<span><span class="va">genes_coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/features.html">genes</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span>, filter <span class="op">=</span> <span class="fu">GeneNameFilter</span><span class="op">(</span><span class="va">genes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">genes_coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">genes_coord</span><span class="op">)</span></span>
<span><span class="va">genes_coord</span> <span class="op">&lt;-</span> <span class="va">genes_coord</span><span class="op">[</span><span class="va">genes_coord</span><span class="op">$</span><span class="va">seqnames</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">genes_coord</span> <span class="op">&lt;-</span> <span class="va">genes_coord</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqnames"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"gene_name"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">genes_coord</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHROM"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"id"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">genes_coord</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
</div>
</section><section class="level2" data-number="3.3"><h3 data-number="3.3" id="match-cells-and-features">
<span class="header-section-number">3.3</span> Match cells and features<a class="anchor" aria-label="anchor" href="#match-cells-and-features"></a>
</h3>
<div>
<blockquote>
<p><strong>Important</strong></p>
<p>Cell names (columns) and features names (rows) must match between assays and features coordinates table ids.</p>
</blockquote>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make sure row names of mat_counts match features ids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat_counts_ATAC</span><span class="op">)</span>, <span class="va">peaks_coord</span><span class="op">[</span>, <span class="st">"id"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat_counts_ATAC</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">peaks_coord</span><span class="op">[</span>, <span class="st">"id"</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat_counts_RNA</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">genes_coord</span><span class="op">[</span>, <span class="st">"id"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Cell names format must match between different assays</span></span>
<span><span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat_counts_ATAC</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat_counts_RNA</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</section></section><section class="section level2" data-number="4"><h2 data-number="4" id="bulk-log-ratios">
<span class="header-section-number">4</span> Bulk log-ratios<a class="anchor" aria-label="anchor" href="#bulk-log-ratios"></a>
</h2>
<p>A table of log ratios computed from matched bulk sequencing (e.g. Whole Genome Sequencing) can optionally be provided to <code><a href="../reference/CreateMuscadetObject.html">CreateMuscadetObject()</a></code> and will be displayed at the bottom of log ratio heatmaps for validation purposes (<a href="#tbl-bulk-lrr" class="quarto-xref">Table 6</a>).</p>
<p>Log ratios for bulk sequencing can be obtained through <a href="https://github.com/mskcc/facets" class="external-link">FACETS</a><a class="footnote-ref" role="doc-noteref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Shen R, Seshan VE. &lt;strong&gt;FACETS: allele-specific copy number and clonal heterogeneity analysis tool for high-throughput DNA sequencing. &lt;/strong&gt; &lt;em&gt;Nucleic Acids Res&lt;/em&gt; (2016). &lt;a href="https://www.doi.org/10.1093/nar/gkw520" class="external-link"&gt;https://www.doi.org/10.1093/nar/gkw520&lt;/a&gt;&lt;/p&gt;'><sup>3</sup></a> analysis.</p>
<p>See <code><a href="../reference/bulk_lrr.html">?bulk_lrr</a></code> for full documentation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"bulk_lrr"</span><span class="op">)</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bulk_lrr</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div id="tbl-bulk-lrr" class="cell quarto-float quarto-figure quarto-figure-center">
<figure class="quarto-float quarto-float-tbl"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-bulk-lrr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table 6: Bulk log ratios table example
</figcaption><div aria-describedby="tbl-bulk-lrr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="table do-not-create-environment cell caption-top">
<thead><tr class="header">
<th style="text-align: right;">CHROM</th>
<th style="text-align: right;">start</th>
<th style="text-align: right;">end</th>
<th style="text-align: right;">lrr</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">16100</td>
<td style="text-align: right;">78558400</td>
<td style="text-align: right;">-0.1742290</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">78558700</td>
<td style="text-align: right;">80533900</td>
<td style="text-align: right;">-0.8500686</td>
</tr>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">80534279</td>
<td style="text-align: right;">80603200</td>
<td style="text-align: right;">-0.1721919</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">80603500</td>
<td style="text-align: right;">99964900</td>
<td style="text-align: right;">-0.8668536</td>
</tr>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">99965200</td>
<td style="text-align: right;">122514900</td>
<td style="text-align: right;">-0.1599042</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">122520400</td>
<td style="text-align: right;">124588600</td>
<td style="text-align: right;">0.0511974</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
            // default icon
            link.classList.add("external");
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marie Denoulet, Nils Giordano, Eric Letouzé, Céline Vallot, Stéphane Minvielle.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
