<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Multiomic copy number analysis tutorial • muscadet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multiomic copy number analysis tutorial">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">muscadet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/muscadet.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/data-prep.html">Preparation of input data</a></li>
    <li><a class="dropdown-item" href="../articles/tutorial_multiome.html">Multiomic copy number analysis tutorial</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ICAGEN/muscadet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Multiomic copy number analysis tutorial</h1>
            
            <h4 data-toc-skip class="date">January 21, 2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ICAGEN/muscadet/blob/v0.2.1/vignettes/articles/tutorial_multiome.Rmd" class="external-link"><code>vignettes/articles/tutorial_multiome.Rmd</code></a></small>
      <div class="d-none name"><code>tutorial_multiome.Rmd</code></div>
    </div>

    
    
<p>This document illustrates a <strong>complete analysis
workflow</strong>, provided as a <strong>step-by-step tutorial</strong>,
using the <code>muscadet</code> package on <strong>10x Genomics Multiome
(scATAC-seq + scRNA-seq) single cell data</strong> from multiple myeloma
tumor cells, together with matched normal reference cells.</p>
<p>The goals of this tutorial are to:</p>
<ul>
<li>Construct a multi-omic <code>muscadet</code> object</li>
<li>Compute genome-wide log R ratio (coverage) profiles for ATAC and RNA
modalities</li>
<li>Perform integrated clustering of cells based on copy-number-driven
coverage signals</li>
<li>Select a biologically meaningful clustering partition</li>
<li>Call copy number alterations (CNAs) at the cluster level</li>
<li>Visualize CNA profiles and relate them to ATAC- and RNA-based UMAP
embeddings</li>
</ul>
<hr>
<div class="section level2">
<h2 id="data-description">Data description<a class="anchor" aria-label="anchor" href="#data-description"></a>
</h2>
<blockquote>
<p><strong>Note</strong></p>
<p>The data used in this analysis are not publicly available; however,
the example dataset included in the <code>muscadet</code> package can be
used instead to test functionalities, in that case please refer to the
<a href="https://icagen.github.io/muscadet/articles/muscadet.html">Get
Started vignette</a> for guidance.</p>
</blockquote>
<p>The dataset used in this tutorial consists of:</p>
<ul>
<li>
<strong>Tumor cells</strong> from a multiple myeloma sample profiled
using 10x Genomics Multiome technology. Normal cells have already been
removed from the tumor dataset prior to this analysis</li>
<li>
<strong>Reference (normal) cells</strong>, used to normalize
coverage and compute log R ratio profiles. These reference cells
originate from individuals distinct from the tumor sample</li>
<li>
<strong>Count matrices</strong> from both scATAC-seq and scRNA-seq
modalities</li>
<li>
<strong>Allele-specific counts</strong> from both scATAC-seq and
scRNA-seq modalities, required for downstream allele-specific CNA
calling</li>
<li>A bulk whole-genome sequencing (WGS) log R ratio profile, used as an
external reference to validate single-cell coverage profiles</li>
</ul>
<p>Load the <code>muscadet</code> package. Additionally, this tutorial
uses the following packages: <code>clustree</code>, <code>dplyr</code>,
and <code>ggplot2</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ICAGEN/muscadet" class="external-link">muscadet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lazappi/clustree" class="external-link">clustree</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>The following inputs are loaded:</p>
<ul>
<li>
<strong>Count matrices</strong> (<code>mat_counts_*</code>) for ATAC
and RNA, tumor and reference cells</li>
<li>
<strong>Allele count tables</strong> (<code>allele_counts_*</code>)
for ATAC and RNA, tumor and reference cells</li>
<li>
<strong>Genomic coordinates</strong> of features: peaks for ATAC and
genes for RNA</li>
<li>Bulk WGS log R ratio (LRR) values <em>(optional)</em>
</li>
</ul>
<blockquote>
<p><strong>Note</strong></p>
<p>For additional details on input formats and preprocessing steps, see
the <a href="https://icagen.github.io/muscadet/articles/data-prep.html">Preparation
of input data vignette</a>.</p>
</blockquote>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Matrices of raw counts</span></span>
<span><span class="va">mat_counts_atac_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"mat_counts_atac_tumor.Rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mat_counts_rna_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"mat_counts_rna_tumor.Rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mat_counts_atac_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"mat_counts_atac_ref.Rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mat_counts_rna_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"mat_counts_rna_ref.Rds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Data frames of allele-specific counts</span></span>
<span><span class="va">allele_counts_atac_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"allele_counts_atac_tumor.Rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">allele_counts_rna_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"allele_counts_rna_tumor.Rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">allele_counts_atac_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"allele_counts_atac_ref.Rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">allele_counts_rna_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"allele_counts_rna_ref.Rds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Genomic coordinates</span></span>
<span><span class="va">peaks_coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"peaks_coord.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">genes_coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"genes_coord.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Coverage (log R ratio) from matched WGS data</span></span>
<span><span class="va">wgs_lrr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"wgs_lrr.tsv"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect dimensions of count matrices</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    ATAC_tumor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat_counts_atac_tumor</span><span class="op">)</span>,</span>
<span>    RNA_tumor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat_counts_rna_tumor</span><span class="op">)</span>,</span>
<span>    ATAC_ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat_counts_atac_ref</span><span class="op">)</span>,</span>
<span>    RNA_ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat_counts_rna_ref</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cells"</span>, <span class="st">"Features"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Cells</th>
<th align="right">Features</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ATAC_tumor</td>
<td align="right">2484</td>
<td align="right">239968</td>
</tr>
<tr class="even">
<td align="left">RNA_tumor</td>
<td align="right">2445</td>
<td align="right">21360</td>
</tr>
<tr class="odd">
<td align="left">ATAC_ref</td>
<td align="right">22175</td>
<td align="right">239968</td>
</tr>
<tr class="even">
<td align="left">RNA_ref</td>
<td align="right">19526</td>
<td align="right">23394</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect number of unique variant positions and cells</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    ATAC_tumor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">allele_counts_atac_tumor</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">allele_counts_atac_tumor</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    RNA_tumor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">allele_counts_rna_tumor</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">allele_counts_rna_tumor</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    ATAC_ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">allele_counts_atac_ref</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">allele_counts_atac_ref</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    RNA_ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">allele_counts_rna_ref</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">allele_counts_rna_ref</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cells"</span>, <span class="st">"Variant positions"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Cells</th>
<th align="right">Variant positions</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ATAC_tumor</td>
<td align="right">2620</td>
<td align="right">1127639</td>
</tr>
<tr class="even">
<td align="left">RNA_tumor</td>
<td align="right">2616</td>
<td align="right">127359</td>
</tr>
<tr class="odd">
<td align="left">ATAC_ref</td>
<td align="right">25865</td>
<td align="right">1380988</td>
</tr>
<tr class="even">
<td align="left">RNA_ref</td>
<td align="right">25758</td>
<td align="right">220287</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect number of features</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">peaks_coord</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 239968</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">genes_coord</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 20018</span></span></code></pre>
<hr>
</div>
<div class="section level2">
<h2 id="creating-muscadet-objects">Creating muscadet objects<a class="anchor" aria-label="anchor" href="#creating-muscadet-objects"></a>
</h2>
<p>Each omic modality is first wrapped into a <code>muscomic</code>
object, which stores:</p>
<ul>
<li>Raw counts matrices</li>
<li>Genomic feature coordinates</li>
<li>Optional allele counts</li>
</ul>
<p>These modality-specific objects are then combined into a single
<code>muscadet</code> object, which represents a multi-omic single-cell
dataset and serves as the central container for all downstream analyses
in muscadet.</p>
<p>Tumor and reference datasets are created <strong>separately</strong>.
Optional bulk WGS log R ratio data may be added to the tumor
<code>muscadet</code> object, but are not required - and typically not
included - for the reference object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Tumor cells</span></span>
<span><span class="va">omic_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="st">"ATAC"</span>,</span>
<span>    mat_counts <span class="op">=</span> <span class="va">mat_counts_atac_tumor</span>,</span>
<span>    allele_counts <span class="op">=</span> <span class="va">allele_counts_atac_tumor</span>,</span>
<span>    features <span class="op">=</span> <span class="va">peaks_coord</span><span class="op">)</span></span>
<span></span>
<span><span class="va">omic_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>    mat_counts <span class="op">=</span> <span class="va">mat_counts_rna_tumor</span>,</span>
<span>    allele_counts <span class="op">=</span> <span class="va">allele_counts_rna_tumor</span>,</span>
<span>    features <span class="op">=</span> <span class="va">genes_coord</span><span class="op">)</span></span>
<span></span>
<span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateMuscadetObject.html">CreateMuscadetObject</a></span><span class="op">(</span></span>
<span>    omics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">omic_atac</span>, <span class="va">omic_rna</span><span class="op">)</span>,</span>
<span>    bulk.lrr <span class="op">=</span> <span class="va">wgs_lrr</span>,</span>
<span>    bulk.label <span class="op">=</span> <span class="st">"WGS"</span>,</span>
<span>    genome <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Reference cells</span></span>
<span><span class="va">omic_atac_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="st">"ATAC"</span>,</span>
<span>    mat_counts <span class="op">=</span> <span class="va">mat_counts_atac_ref</span>,</span>
<span>    allele_counts <span class="op">=</span> <span class="va">allele_counts_atac_ref</span>,</span>
<span>    features <span class="op">=</span> <span class="va">peaks_coord</span><span class="op">)</span></span>
<span></span>
<span><span class="va">omic_rna_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateMuscomicObject.html">CreateMuscomicObject</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>    mat_counts <span class="op">=</span> <span class="va">mat_counts_rna_ref</span>,</span>
<span>    allele_counts <span class="op">=</span> <span class="va">allele_counts_rna_ref</span>,</span>
<span>    features <span class="op">=</span> <span class="va">genes_coord</span><span class="op">)</span></span>
<span></span>
<span><span class="va">muscadet_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateMuscadetObject.html">CreateMuscadetObject</a></span><span class="op">(</span></span>
<span>    omics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">omic_atac_ref</span>, <span class="va">omic_rna_ref</span><span class="op">)</span>,</span>
<span>    genome <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p><strong>Important</strong></p>
<p><strong>Allele counts are optional for clustering, but required for
downstream CNA calling</strong>. They can be provided either at object
creation or added later using the <code><a href="../reference/addAlleleCounts.html">addAlleleCounts()</a></code>. As a
result, cells can be clustered even when allele count data are not yet
available. This approach can be useful when clustering and exploratory
analyses are performed first, and CNA calling is deferred to a later
step.</p>
</blockquote>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># To add allele counts later in the analysis</span></span>
<span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addAlleleCounts.html">addAlleleCounts</a></span><span class="op">(</span></span>
<span>    <span class="va">muscadet</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">allele_counts_atac_tumor</span>, <span class="va">allele_counts_rna_tumor</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">muscadet_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addAlleleCounts.html">addAlleleCounts</a></span><span class="op">(</span></span>
<span>    <span class="va">muscadet_ref</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">allele_counts_atac_ref</span>, <span class="va">allele_counts_rna_ref</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
<p>The content of the <code>muscadet</code> objects can be seen:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect objects</span></span>
<span><span class="va">muscadet</span></span></code></pre></div>
<pre><code><span><span class="co">## A muscadet object </span></span>
<span><span class="co">##  2 omics: ATAC, RNA </span></span>
<span><span class="co">##  types: ATAC, RNA </span></span>
<span><span class="co">##  labels: scATAC-seq, scRNA-seq </span></span>
<span><span class="co">##  cells: 2484, 2445 (common: 2309, total: 2620) </span></span>
<span><span class="co">##  counts: 2484 cells x 239968 features (peaks), 2445 cells x 14872 features (genes) </span></span>
<span><span class="co">##  logratio: None</span></span>
<span><span class="co">##  variant positions: 1127639, 127359 </span></span>
<span><span class="co">##  bulk data: WGS </span></span>
<span><span class="co">##  clustering: None </span></span>
<span><span class="co">##  CNA calling: None </span></span>
<span><span class="co">##  genome: hg38</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">muscadet_ref</span></span></code></pre></div>
<pre><code><span><span class="co">## A muscadet object </span></span>
<span><span class="co">##  2 omics: ATAC, RNA </span></span>
<span><span class="co">##  types: ATAC, RNA </span></span>
<span><span class="co">##  labels: scATAC-seq, scRNA-seq </span></span>
<span><span class="co">##  cells: 22175, 19526 (common: 15836, total: 25865) </span></span>
<span><span class="co">##  counts: 22175 cells x 239968 features (peaks), 19526 cells x 15585 features (genes) </span></span>
<span><span class="co">##  logratio: None</span></span>
<span><span class="co">##  variant positions: 1380988, 220287 </span></span>
<span><span class="co">##  bulk data: None </span></span>
<span><span class="co">##  clustering: None </span></span>
<span><span class="co">##  CNA calling: None </span></span>
<span><span class="co">##  genome: hg38</span></span></code></pre>
<p>The <code>muscadet</code> class of objects contains 5 slots:</p>
<ul>
<li>
<code>omics</code>: a list of muscomic objects, one per omic
modality</li>
<li>
<code>bulk.data</code>: log R ratio data derived from matched bulk
sequencing, when available</li>
<li>
<code>clustering</code>: storage of clustering results and
associated metrics</li>
<li>
<code>cnacalling</code>: results of CNA calling</li>
<li>
<code>genome</code>: reference genome information used throughout
the analysis</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">muscadet</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Formal class 'muscadet' [package "muscadet"] with 5 slots</span></span>
<span><span class="co">##   ..@ omics     :List of 2</span></span>
<span><span class="co">##   ..@ bulk.data :List of 2</span></span>
<span><span class="co">##   ..@ clustering: list()</span></span>
<span><span class="co">##   ..@ cnacalling: list()</span></span>
<span><span class="co">##   ..@ genome    : chr "hg38"</span></span></code></pre>
<p>To access information in <code>muscomic</code> and
<code>muscadet</code> objects several functions can be used:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get feature coordinates data frame</span></span>
<span><span class="fu"><a href="../reference/muscadet-methods.html">coordFeatures</a></span><span class="op">(</span><span class="va">omic_atac</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/muscadet-methods.html">coordFeatures</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">$</span><span class="va">ATAC</span></span>
<span></span>
<span><span class="co"># Get count matrices and log R ratio matrices</span></span>
<span><span class="fu"><a href="../reference/muscadet-methods.html">matCounts</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">$</span><span class="va">ATAC</span></span>
<span><span class="fu"><a href="../reference/muscadet-methods.html">matLogRatio</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">$</span><span class="va">ATAC</span></span>
<span></span>
<span><span class="co"># Get name of cells and features</span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Cells.html" class="external-link">Cells</a></span><span class="op">(</span><span class="va">omic_atac</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Cells.html" class="external-link">Features</a></span><span class="op">(</span><span class="va">omic_atac</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">union</span>, <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Cells.html" class="external-link">Cells</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">)</span> <span class="co"># all cells from ATAC and RNA omics</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="computing-log-r-ratios">Computing log R ratios<a class="anchor" aria-label="anchor" href="#computing-log-r-ratios"></a>
</h2>
<p>Genome-wide log R ratios are computed independently for each omic
using <code><a href="../reference/computeLogRatio.html">computeLogRatio()</a></code>.</p>
<p>Tumor coverage is normalized against reference cells, the result is a
genome-ordered matrix where deviations from zero reflect copy-number
gains or losses. These resulting log R ratio matrices are stored
directly within the <code>muscadet</code> objects returned by the
function.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute log R ratios for ATAC</span></span>
<span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeLogRatio.html">computeLogRatio</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">muscadet</span>,</span>
<span>    reference <span class="op">=</span> <span class="va">muscadet_ref</span>,</span>
<span>    omic <span class="op">=</span> <span class="st">"ATAC"</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"ATAC"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># compute log R ratios for RNA</span></span>
<span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeLogRatio.html">computeLogRatio</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">muscadet</span>,</span>
<span>    reference <span class="op">=</span> <span class="va">muscadet_ref</span>,</span>
<span>    omic <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">muscadet</span></span></code></pre></div>
<pre><code><span><span class="co">## A muscadet object </span></span>
<span><span class="co">##  2 omics: ATAC, RNA </span></span>
<span><span class="co">##  types: ATAC, RNA </span></span>
<span><span class="co">##  labels: scATAC-seq, scRNA-seq </span></span>
<span><span class="co">##  cells: 2484, 2445 (common: 2309, total: 2620) </span></span>
<span><span class="co">##  counts: 2484 cells x 239968 features (peaks), 2445 cells x 14872 features (genes) </span></span>
<span><span class="co">##  logratio: 2484 cells x 1364 features (windows of peaks), 2445 cells x 11281 features (genes) </span></span>
<span><span class="co">##  variant positions: 1127639, 127359 </span></span>
<span><span class="co">##  bulk data: WGS </span></span>
<span><span class="co">##  clustering: None </span></span>
<span><span class="co">##  CNA calling: None </span></span>
<span><span class="co">##  genome: hg38</span></span></code></pre>
<p>ATAC and RNA are handled using modality-specific normalization
strategies in <code><a href="../reference/computeLogRatio.html">computeLogRatio()</a></code>, which calls for the
<code><a href="../reference/computeLogRatioATAC.html">computeLogRatioATAC()</a></code> and <code><a href="../reference/computeLogRatioRNA.html">computeLogRatioRNA()</a></code>
functions respectively.</p>
<blockquote>
<p><strong>Note</strong></p>
<p>For the ATAC method, the feature name becomes “windows of peaks”,
corresponding to genomic windows derived from peaks, as counts are
aggregated into bins along the genome according to the
<code>windowSize</code> and <code>slidingSize</code> parameters.</p>
<p>For the RNA method, log R ratio values are smoothed using a running
mean over windows defined by the <code>genesPerWindow</code>
parameter.</p>
</blockquote>
<p>The default feature-filtering parameters for the ATAC workflow
implemented in <code><a href="../reference/computeLogRatioATAC.html">computeLogRatioATAC()</a></code> are:</p>
<ul>
<li>A minimum read average per window in reference cells
(<code>minReads</code>) set to 5</li>
<li>A minimum number of peaks per window (<code>minPeaks</code>) set to
100</li>
</ul>
<p>The default feature-filtering parameters for the RNA workflow
implemented in <code><a href="../reference/computeLogRatioRNA.html">computeLogRatioRNA()</a></code> are:</p>
<ul>
<li>A minimum of reads in reference cells (<code>refReads</code>) set to
100</li>
<li>A minimum of average reads per reference cell
(<code>refMeanReads</code>) set to 0.01</li>
</ul>
<blockquote>
<p><strong>Important</strong></p>
<p>These default feature-filtering values are tailored to this dataset
and may not be appropriate for other datasets.</p>
</blockquote>
<p>To customize or refine the feature filtering applied by
<code><a href="../reference/computeLogRatio.html">computeLogRatio()</a></code>, users are encouraged to inspect the
underlying data distributions and to examine the filtering status of
individual features.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">ATAC_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/muscadet-methods.html">coordFeatures</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">$</span><span class="va">ATAC</span></span>
<span></span>
<span><span class="va">ATAC_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">ATAC_features</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">nPeaks</span>, y <span class="op">=</span> <span class="va">meanReads.ref</span>, color <span class="op">=</span> <span class="va">keep</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">100</span> , linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># minPeaks threshold</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">5</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># minReads threshold</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Number of peaks (minPeaks)"</span>, y <span class="op">=</span> <span class="st">"Mean of reads in reference cells (minReads)"</span>,</span>
<span>         title <span class="op">=</span> <span class="st">"ATAC features (windows of peaks) filtered by thresholds"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"figures/ATAC_features.png"</span>, <span class="va">ATAC_plot</span><span class="op">)</span></span>
<span></span>
<span><span class="va">RNA_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/muscadet-methods.html">coordFeatures</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">$</span><span class="va">RNA</span></span>
<span></span>
<span><span class="va">RNA_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">RNA_features</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sumReads.ref</span>, y <span class="op">=</span> <span class="va">meanReads.ref</span>, color <span class="op">=</span> <span class="va">keep</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">100</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># refReads threshold</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0.01</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># refMeanReads threshold</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Sum of reads in reference cells (refReads)"</span>, y <span class="op">=</span> <span class="st">"Mean of reads in reference cells (refMeanReads)"</span>,</span>
<span>         title <span class="op">=</span> <span class="st">"RNA features (genes) filtered by thresholds"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"figures/RNA_features.png"</span>, <span class="va">RNA_plot</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/ATAC_features.png" style="width:70.0%" fig-align="center" alt="ATAC features"><img src="figures/RNA_features.png" style="width:70.0%" fig-align="center" alt="RNA features"></p>
<hr>
</div>
<div class="section level2">
<h2 id="multi-omic-clustering">Multi-omic clustering<a class="anchor" aria-label="anchor" href="#multi-omic-clustering"></a>
</h2>
<p>Cells are clustered based on their log R ratio profiles using the
<code><a href="../reference/clusterMuscadet.html">clusterMuscadet()</a></code> function. Two clustering strategies are
currently available:</p>
<ul>
<li><p><strong><code>method = "seurat"</code></strong>: Graph-based
clustering using the <a href="https://satijalab.org/seurat/" class="external-link"><code>Seurat</code></a> package.
This approach constructs a nearest-neighbor graph from a weighted
combination of multiple modalities (Weighted Nearest-Neighbor graph,
WNN), using selected principal components from each modality, followed
by community detection to identify clusters (see
<code><a href="../reference/cluster_seurat.html">cluster_seurat()</a></code>).</p></li>
<li><p><strong><code>method = "hclust"</code></strong>: Multi-omic
integration via Similarity Network Fusion (SNF), followed by
hierarchical clustering of the fused similarity matrix to identify cell
clusters (see <code><a href="../reference/cluster_hclust.html">cluster_hclust()</a></code>).</p></li>
</ul>
<p>For these data, clustering is performed using the
<strong>Seurat-based approach</strong> (the default method). This
approach enables joint neighborhood construction while preserving
omic-specific signal.</p>
<p>In this framework, each omic is first processed independently to
obtain a low-dimensional representation that captures the dominant
sources of variation in the data (copy-number–driven variation reflected
in log R ratios). From these representations, k nearest neighbor graphs
are constructed separately for each omic.</p>
<p>The omic-specific neighbor graphs are then integrated into a unified
graph, which reflects shared cell–cell relationships across modalities
while remaining robust to modality-specific noise or missing
information. This integrated graph serves as the basis for
clustering.</p>
<p>Clustering is performed using a graph-based community detection
algorithm, and multiple clustering resolutions are explored
simultaneously via the <code>res_range</code> argument. This allows the
user to evaluate partitions at different levels of granularity without
rerunning the entire clustering step.</p>
<p>The resulting cluster assignments are stored in the
<code>muscadet</code> object as a named list, with one element per
resolution (<code>muscadet$clustering$clusters</code>).</p>
<p>Printing clusters sizes tables at this stage is helpful to view the
number of cells per cluster per partitions.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clustering</span></span>
<span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterMuscadet.html">clusterMuscadet</a></span><span class="op">(</span></span>
<span>    <span class="va">muscadet</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"seurat"</span>,</span>
<span>    res_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    dims_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 9</span></span>
<span><span class="co">##  $ params       :List of 9</span></span>
<span><span class="co">##  $ pcs          :List of 2</span></span>
<span><span class="co">##  $ nn           :Formal class 'Neighbor' [package "SeuratObject"] with 5 slots</span></span>
<span><span class="co">##  $ graph        :Formal class 'Graph' [package "SeuratObject"] with 7 slots</span></span>
<span><span class="co">##  $ dist         : num [1:2309, 1:2309] -0.182 0.818 0.818 0.818 0.818 ...</span></span>
<span><span class="co">##   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##  $ umap         : num [1:2309, 1:2] 0.43 0.702 -3.084 3.103 3.065 ...</span></span>
<span><span class="co">##   ..- attr(*, "scaled:center")= num [1:2] 1.328 0.722</span></span>
<span><span class="co">##   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##  $ clusters     :List of 5</span></span>
<span><span class="co">##  $ silhouette   :List of 3</span></span>
<span><span class="co">##  $ partition.opt: chr "0.3"</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clusters</span>, <span class="va">table</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $`0.1`</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    1    2    3 </span></span>
<span><span class="co">## 1156  776  688 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`0.2`</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    1    2    3 </span></span>
<span><span class="co">## 1156  780  684 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`0.3`</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    1    2    3    4    5 </span></span>
<span><span class="co">## 1011  776  359  332  142 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`0.4`</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    1    2    3    4    5 </span></span>
<span><span class="co">## 1009  780  355  332  144 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`0.5`</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    1    2    3    4    5 </span></span>
<span><span class="co">## 1005  773  362  334  146</span></span></code></pre>
<blockquote>
<p><strong>Important</strong></p>
<p>When performing Seurat-based clustering, the <code>dims_list</code>
parameter specifies which dimensions of the low-dimensional embeddings
(PCA) are used for neighbor graph construction. Selecting an appropriate
number of dimensions is important: too few may miss relevant biological
variation, while too many can introduce noise. In practice, dimensions
capturing the majority of variance in the data are used, and this can be
adjusted depending on dataset size and complexity.</p>
</blockquote>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract variance info</span></span>
<span><span class="va">get_variance_df</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">pc_obj</span>, <span class="va">label</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>        PC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">pc_obj</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="st">"Proportion of Variance"</span>, <span class="op">]</span><span class="op">)</span>,</span>
<span>        Proportion <span class="op">=</span> <span class="va">pc_obj</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="st">"Proportion of Variance"</span>, <span class="op">]</span>,</span>
<span>        Cumulative <span class="op">=</span> <span class="va">pc_obj</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="st">"Cumulative Proportion"</span>, <span class="op">]</span>,</span>
<span>        PCA <span class="op">=</span> <span class="va">label</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">df_atac</span> <span class="op">&lt;-</span> <span class="fu">get_variance_df</span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">pcs</span><span class="op">$</span><span class="va">ATAC</span>, <span class="st">"ATAC"</span><span class="op">)</span></span>
<span><span class="va">df_rna</span> <span class="op">&lt;-</span> <span class="fu">get_variance_df</span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">pcs</span><span class="op">$</span><span class="va">RNA</span>, <span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="va">combined_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">df_atac</span>, <span class="va">df_rna</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Dimensions used for this dataset</span></span>
<span><span class="va">dims_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">thresholds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    PC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">dims_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">dims_list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    PCA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">omics</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot 20 first PCs</span></span>
<span><span class="va">PCs_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">combined_df</span>, <span class="va">PC</span> <span class="op">&lt;=</span> <span class="fl">20</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC</span>, fill <span class="op">=</span> <span class="va">PCA</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">Proportion</span><span class="op">)</span>, position <span class="op">=</span> <span class="st">"dodge"</span>, width <span class="op">=</span> <span class="fl">0.7</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">Cumulative</span>, color <span class="op">=</span> <span class="va">PCA</span>, group <span class="op">=</span> <span class="va">PCA</span><span class="op">)</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">Cumulative</span>, color <span class="op">=</span> <span class="va">PCA</span>, group <span class="op">=</span> <span class="va">PCA</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span></span>
<span>        labels <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="va">x</span><span class="op">)</span>, <span class="st">"%"</span><span class="op">)</span>,</span>
<span>        expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="st">"Variance Explained by Principal Components"</span>,</span>
<span>        x <span class="op">=</span> <span class="st">"Principal Component"</span>,</span>
<span>        y <span class="op">=</span> <span class="st">"Variance Explained"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">thresholds</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">PC</span>, color <span class="op">=</span> <span class="va">PCA</span><span class="op">)</span>,</span>
<span>        linetype <span class="op">=</span> <span class="st">"dashed"</span>,</span>
<span>        show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"figures/PCs_variance.png"</span>, <span class="va">PCs_plot</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="figures/PCs_variance.png" style="width:70.0%" fig-align="center" alt="Variance Explained by Principal Components"><div class="figcaption">Variance Explained by Principal Components</div>
</div>
<hr>
</div>
<div class="section level2">
<h2 id="exploring-clustering-partitions">Exploring clustering partitions<a class="anchor" aria-label="anchor" href="#exploring-clustering-partitions"></a>
</h2>
<div class="section level3">
<h3 id="clustering-stability-with-clustree">Clustering stability with clustree<a class="anchor" aria-label="anchor" href="#clustering-stability-with-clustree"></a>
</h3>
<p>To explore how clustering structure evolves across resolutions, we
use the <code><a href="https://lazappi.github.io/clustree/reference/clustree.html" class="external-link">clustree()</a></code> visualization provided by the <a href="https://lazappi.github.io/clustree/index.html" class="external-link"><code>clustree</code></a>
package. This representation summarizes the relationships between
clustering partitions by tracking how clusters split, merge, or persist
as the resolution parameter increases.</p>
<p>In a clustree plot, nodes represent clusters at a given resolution,
while edges indicate how cells are redistributed between clusters when
moving to a higher resolution. This makes it possible to visually assess
the stability of cluster assignments across resolutions.</p>
<p>This step helps to:</p>
<ul>
<li>Detect resolutions that are overly granular, characterized by
excessive fragmentation or the appearance of very small clusters</li>
<li>Identify stable cluster structures that persist across multiple
consecutive resolutions, suggesting robust biological signal</li>
<li>Select a resolution that provides an appropriate balance between
interpretability and biological relevance</li>
</ul>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clustering visualization - clustree</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lazappi/clustree" class="external-link">clustree</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract common cells across omics</span></span>
<span><span class="co"># (clustering is performed on cells common to all omic modalities,</span></span>
<span><span class="co"># cells missing one modality are imputed afterward and included in the output)</span></span>
<span><span class="va">common_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>, <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Cells.html" class="external-link">Cells</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Construct clustree input</span></span>
<span><span class="va">partitions</span> <span class="op">&lt;-</span> <span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clusters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span>.id <span class="op">=</span> <span class="st">"res"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">common_cells</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">partitions</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"res_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot clustering tree</span></span>
<span><span class="va">clustree_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lazappi.github.io/clustree/reference/clustree.html" class="external-link">clustree</a></span><span class="op">(</span><span class="va">partitions</span>, prefix <span class="op">=</span> <span class="st">"res_"</span>, node_label <span class="op">=</span> <span class="st">"size"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>size <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"figures/clustree.png"</span>, <span class="va">clustree_plot</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="figures/clustree.png" style="width:80.0%" fig-align="center" alt="Clustree plot"><div class="figcaption">Clustree plot</div>
</div>
</div>
<div class="section level3">
<h3 id="umap-visualization">UMAP visualization<a class="anchor" aria-label="anchor" href="#umap-visualization"></a>
</h3>
<p>UMAP embeddings provide an intuitive two-dimensional representation
of the clustering space. They are particularly useful for visualizing
overall cell relationships and the relative positioning of clusters.</p>
<p>It is important to note that UMAP is used solely for visualization:
clusters are defined in the original high-dimensional space, and UMAP
does not influence the clustering itself.</p>
<p>Concordance between UMAP separation and observed CNA patterns can be
reassuring and help validate cluster assignments, but discrepancies do
not necessarily indicate problems, as some biologically meaningful
structure may not be fully captured in two dimensions.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># clustering visualization - silhouette widths</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">p</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"figures"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"UMAP_res_"</span>, <span class="va">p</span>, <span class="st">".png"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">title</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"method="</span>, <span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">params</span><span class="op">[[</span><span class="st">"method"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"omics="</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">params</span><span class="op">[[</span><span class="st">"omics"</span><span class="op">]</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"dims="</span>, <span class="st">"8-8"</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"res="</span>, <span class="va">p</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="../reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span><span class="va">muscadet</span>, partition <span class="op">=</span> <span class="va">p</span>, title <span class="op">=</span> <span class="va">title</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="va">filename</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="float">
<img src="figures/UMAP_res_0.1.png" style="width:60.0%" fig-align="center" alt="UMAP plot - res 0.1"><div class="figcaption">UMAP plot - res 0.1</div>
</div>
<div class="float">
<img src="figures/UMAP_res_0.2.png" style="width:60.0%" fig-align="center" alt="UMAP plot - res 0.2"><div class="figcaption">UMAP plot - res 0.2</div>
</div>
<div class="float">
<img src="figures/UMAP_res_0.3.png" style="width:60.0%" fig-align="center" alt="UMAP plot - res 0.3"><div class="figcaption">UMAP plot - res 0.3</div>
</div>
<div class="float">
<img src="figures/UMAP_res_0.4.png" style="width:60.0%" fig-align="center" alt="UMAP plot - res 0.4"><div class="figcaption">UMAP plot - res 0.4</div>
</div>
<div class="float">
<img src="figures/UMAP_res_0.5.png" style="width:60.0%" fig-align="center" alt="UMAP plot - res 0.5"><div class="figcaption">UMAP plot - res 0.5</div>
</div>
</div>
<div class="section level3">
<h3 id="genome-wide-heatmaps-of-coverage-profiles">Genome-wide heatmaps of coverage profiles<a class="anchor" aria-label="anchor" href="#genome-wide-heatmaps-of-coverage-profiles"></a>
</h3>
<p>Genome-wide heatmaps provide a direct and intuitive visualization of
copy-number–driven signal across single cells and cell populations. In
<code>muscadet</code>, genome-wide coverage or log R ratio profiles can
be visualized as heatmaps using the <code><a href="../reference/heatmapMuscadet.html">heatmapMuscadet()</a></code>
function, based on a selected clustering partition stored in the
<code>muscadet</code> object.</p>
<p>At the single-cell level, heatmaps display genome-wide log R ratio
profiles for individual cells, ordered by their cluster assignments.
This representation makes it possible to assess both intra-cluster
homogeneity and inter-cluster differences in copy-number profiles.</p>
<p>In addition, <code>muscadet</code> allows the visualization of
aggregated heatmaps, in which log R ratio values are averaged across
cells within each cluster. These cluster-level summaries provide an
overview of recurrent copy-number alteration (CNA) patterns and can
facilitate the comparison of subclonal populations.</p>
<p>Together, these visualizations provide two complementary
perspectives:</p>
<ul>
<li>Single-cell heatmaps, ordered by clusters, which highlight
cell-to-cell variability and cluster coherence</li>
<li>Cluster-average heatmaps, which summarize dominant CNA patterns and
emphasize subclonal structure</li>
</ul>
<p>When interpreting genome-wide heatmaps, the following points are
particularly informative:</p>
<ul>
<li>Vertical shifts spanning contiguous genomic regions typically
reflect chromosome-scale or arm-level copy number gains or losses</li>
<li>Consistent log R ratio patterns across cells within the same cluster
support the biological relevance and robustness of the inferred
clusters</li>
</ul>
<p>These heatmaps are therefore a key visualization step for validating
clustering results and for guiding downstream pseudo-bulk aggregation
and CNA calling.</p>
<blockquote>
<p><strong>Note</strong></p>
<p>Large figures in this tutorial, such as heatmaps, are generated at
lower resolution by default to reduce storage requirements; users may
increase the resolution (or simply omit the argument to use the default
value) if higher-quality output is desired.</p>
</blockquote>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># clustering visualization - genome-wide coverage heatmaps</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">p</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"figures"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"heatmap_res_"</span>, <span class="va">p</span>, <span class="st">".png"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">title</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"method="</span>, <span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">params</span><span class="op">[[</span><span class="st">"method"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"omics="</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">params</span><span class="op">[[</span><span class="st">"omics"</span><span class="op">]</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"dims="</span>, <span class="st">"8-8"</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"res="</span>, <span class="va">p</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="../reference/heatmapMuscadet.html">heatmapMuscadet</a></span><span class="op">(</span><span class="va">muscadet</span>,</span>
<span>                    <span class="va">filename</span>,</span>
<span>                    partition <span class="op">=</span> <span class="va">p</span>,</span>
<span>                    title <span class="op">=</span> <span class="va">title</span>,</span>
<span>                    show_missing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    png_res <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="float">
<img src="figures/heatmap_res_0.1.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios - res 0.1"><div class="figcaption">Heatmap of log R ratios - res 0.1</div>
</div>
<div class="float">
<img src="figures/heatmap_res_0.2.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios - res 0.2"><div class="figcaption">Heatmap of log R ratios - res 0.2</div>
</div>
<div class="float">
<img src="figures/heatmap_res_0.3.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios - res 0.3"><div class="figcaption">Heatmap of log R ratios - res 0.3</div>
</div>
<div class="float">
<img src="figures/heatmap_res_0.4.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios - res 0.4"><div class="figcaption">Heatmap of log R ratios - res 0.4</div>
</div>
<div class="float" id="fig-ht">
<img src="figures/heatmap_res_0.5.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios - res 0.5"><div class="figcaption">Heatmap of log R ratios - res 0.5</div>
</div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">p</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"figures"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"heatmap_res_"</span>, <span class="va">p</span>, <span class="st">"_av.png"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">title</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"method="</span>, <span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">params</span><span class="op">[[</span><span class="st">"method"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"omics="</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">params</span><span class="op">[[</span><span class="st">"omics"</span><span class="op">]</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"dims="</span>, <span class="st">"8-8"</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"res="</span>, <span class="va">p</span><span class="op">)</span>,</span>
<span>        <span class="st">"| Averages per cluster"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="../reference/heatmapMuscadet.html">heatmapMuscadet</a></span><span class="op">(</span></span>
<span>        <span class="va">muscadet</span>,</span>
<span>        <span class="va">filename</span>,</span>
<span>        partition <span class="op">=</span> <span class="va">p</span>,</span>
<span>        title <span class="op">=</span> <span class="va">title</span>,</span>
<span>        averages <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="float">
<img src="figures/heatmap_res_0.1_av.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios averages per cluster - res 0.1"><div class="figcaption">Heatmap of log R ratios averages per cluster -
res 0.1</div>
</div>
<div class="float">
<img src="figures/heatmap_res_0.2_av.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios averages per cluster - res 0.2"><div class="figcaption">Heatmap of log R ratios averages per cluster -
res 0.2</div>
</div>
<div class="float">
<img src="figures/heatmap_res_0.3_av.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios averages per cluster - res 0.3"><div class="figcaption">Heatmap of log R ratios averages per cluster -
res 0.3</div>
</div>
<div class="float">
<img src="figures/heatmap_res_0.4_av.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios averages per cluster - res 0.4"><div class="figcaption">Heatmap of log R ratios averages per cluster -
res 0.4</div>
</div>
<div class="float">
<img src="figures/heatmap_res_0.5_av.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios averages per cluster - res 0.5"><div class="figcaption">Heatmap of log R ratios averages per cluster -
res 0.5</div>
</div>
</div>
<div class="section level3">
<h3 id="clustering-quality-metrics">Clustering quality metrics<a class="anchor" aria-label="anchor" href="#clustering-quality-metrics"></a>
</h3>
<p>Silhouette widths are computed automatically during the clustering
step and stored within the <code>muscadet object</code>. These scores
can be visualized using the <code><a href="../reference/plotSil.html">plotSil()</a></code> function.</p>
<p>Silhouette values measure how similar a cell is to its assigned
cluster compared to neighboring clusters. Higher values indicate
better-defined and more clearly separated clusters, whereas low or
negative values suggest ambiguous assignments or overlapping cluster
structure.</p>
<p>When using silhouette widths in practice:</p>
<ul>
<li>Higher silhouette values generally indicate improved cluster
separation</li>
<li>Distributions of silhouette scores should be compared across
clustering resolutions, rather than interpreted in absolute terms for a
single partition</li>
</ul>
<p>Importantly, silhouette-based metrics are intended to complement, not
replace, biological inspection of genome-wide heatmaps and copy number
patterns. In the context of CNA-driven clustering, biological coherence
and interpretability remain the primary criteria for selecting an
appropriate resolution.</p>
<blockquote>
<p><strong>Note</strong></p>
<p>In addition, the <code><a href="../reference/plotIndexes.html">plotIndexes()</a></code> function can be used to
compute and visualize complementary clustering validation indices.</p>
</blockquote>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># clustering visualization - silhouette widths</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">p</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"figures"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"plotSil_res_"</span>, <span class="va">p</span>, <span class="st">".png"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">title</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"method="</span>, <span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">params</span><span class="op">[[</span><span class="st">"method"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"omics="</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">params</span><span class="op">[[</span><span class="st">"omics"</span><span class="op">]</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"dims="</span>, <span class="st">"8-8"</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"res="</span>, <span class="va">p</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="../reference/plotSil.html">plotSil</a></span><span class="op">(</span><span class="va">muscadet</span>, partition <span class="op">=</span> <span class="va">p</span>, title <span class="op">=</span> <span class="va">title</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="va">filename</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="float">
<img src="figures/plotSil_res_0.1.png" style="width:70.0%" fig-align="center" alt="Silhouette plot - res 0.1"><div class="figcaption">Silhouette plot - res 0.1</div>
</div>
<div class="float">
<img src="figures/plotSil_res_0.2.png" style="width:70.0%" fig-align="center" alt="Silhouette plot - res 0.2"><div class="figcaption">Silhouette plot - res 0.2</div>
</div>
<div class="float">
<img src="figures/plotSil_res_0.3.png" style="width:70.0%" fig-align="center" alt="Silhouette plot - res 0.3"><div class="figcaption">Silhouette plot - res 0.3</div>
</div>
<div class="float">
<img src="figures/plotSil_res_0.4.png" style="width:70.0%" fig-align="center" alt="Silhouette plot - res 0.4"><div class="figcaption">Silhouette plot - res 0.4</div>
</div>
<div class="float">
<img src="figures/plotSil_res_0.5.png" style="width:70.0%" fig-align="center" alt="Silhouette plot - res 0.5"><div class="figcaption">Silhouette plot - res 0.5</div>
</div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="cna-calling">CNA calling<a class="anchor" aria-label="anchor" href="#cna-calling"></a>
</h2>
<div class="section level3">
<h3 id="selecting-a-clustering-partition">Selecting a clustering partition<a class="anchor" aria-label="anchor" href="#selecting-a-clustering-partition"></a>
</h3>
<p>Before performing copy number calling, a single clustering partition
must be selected using <code><a href="../reference/assignClusters.html">assignClusters()</a></code>. This partition
defines which cells are aggregated into pseudo-bulk profiles for
downstream analysis.</p>
<blockquote>
<p><strong>Note</strong></p>
<p>In this tutorial, we select the partition at resolution
<code>0.2</code>, which yields three clusters.</p>
<p>At higher resolutions, additional clusters appear: for example,
cluster 5 in the resolution <code>0.3</code> partition corresponds to a
subpopulation with a deletion on chromosome 16q (visible in the
scATAC-seq profile) however, this subpopulation represents less than 6%
of cells, and for clarity and interpretability, it is not included in
downstream analysis for this tutorial.</p>
</blockquote>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assign the chosen clustering partition</span></span>
<span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assignClusters.html">assignClusters</a></span><span class="op">(</span><span class="va">muscadet</span>, partition <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">cnacalling</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    1    2    3 </span></span>
<span><span class="co">## 1156  780  684</span></span></code></pre>
<div class="float">
<img src="figures/heatmap_res_0.2.png" style="width:100.0%" fig-align="center" alt="Heatmap of log R ratios for chosen partition (res 0.2)"><div class="figcaption">Heatmap of log R ratios for chosen partition
(res 0.2)</div>
</div>
</div>
<div class="section level3">
<h3 id="aggregating-counts">Aggregating counts<a class="anchor" aria-label="anchor" href="#aggregating-counts"></a>
</h3>
<p>The <code><a href="../reference/aggregateCounts.html">aggregateCounts()</a></code> function combines counts across
all cells within each cluster and across multiple omics. This produces
pseudo-bulk profiles that are suitable for segmentation and downstream
copy number inference.</p>
<p>This step is mandatory for CNA calling. Aggregation is required to
produce sufficiently robust signal as it improves signal-to-noise ratio:
by combining single-cell measurements, random fluctuations are reduced,
making copy number patterns more apparent.</p>
<p>Reference counts are aggregated similarly across all cells and omics
to remain consistent and comparable.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Merge all counts from all omics from both sample and reference</span></span>
<span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateCounts.html">aggregateCounts</a></span><span class="op">(</span><span class="va">muscadet</span>, <span class="va">muscadet_ref</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="running-cna-calling">Running CNA calling<a class="anchor" aria-label="anchor" href="#running-cna-calling"></a>
</h3>
<p>CNA calling is performed using the <code><a href="../reference/cnaCalling.html">cnaCalling()</a></code>
function, which segments pseudo-bulk profiles and infers copy number
alterations. This function integrates coverage and allelic data to
segment chromosomes and to produce cluster-level copy number calls.</p>
<blockquote>
<p><strong>Note</strong></p>
<p>In this dataset, ATAC-derived coverage profiles exhibit cleaner
signal than RNA-derived profiles. To avoid introducing noise in that
case, CNA calling is restricted to ATAC coverage by setting
<code>omics.coverage = "ATAC"</code>.</p>
</blockquote>
<p>Note that this selection applies only to coverage data; allelic
information can still be used from multiple omics. Coverage from
multiple omics can be used if both are reliable and comparable.</p>
<blockquote>
<p><strong>Important</strong></p>
<p>During CNA calling, a series of filters based on minimum value
thresholds (<code>depthmin.*.*</code>) are applied to remove noise. The
default thresholds are tailored to this dataset, but they may need to be
adjusted for other datasets with different characteristics or sequencing
depth.</p>
</blockquote>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># CNA calling</span></span>
<span><span class="va">muscadet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cnaCalling.html">cnaCalling</a></span><span class="op">(</span><span class="va">muscadet</span>, omics.coverage <span class="op">=</span> <span class="st">"ATAC"</span><span class="op">)</span></span></code></pre></div>
<p>The output is stored within the <code>muscadet</code> object and
includes:</p>
<ul>
<li>Positions (<code>positions</code> /
<code>positions.allcells</code>): base-level data for each genomic
position (coverage/allele counts, log-R ratio, variant allele frequency,
and cluster assignment)</li>
<li>Segments (<code>segments</code> / <code>segments.allcells</code>):
aggregated contiguous positions forming segments and associated data
(median log-R ratio, median variant allele frequency, copy number
estimates, and cluster assignment)</li>
<li>Consensus segments (<code>consensus.segs</code>): unique segments
across clusters (segment coordinates, CNA status, and clonal CNA
status)</li>
<li>Cluster-level summary table (<code>table</code>): consensus segments
annotated per cluster (segment coordinates, cluster assignment, copy
numbers, CNA state (gain/neutral/loss), proportion of cells of cluster,
proportion of total cells, …)</li>
</ul>
</div>
<div class="section level3">
<h3 id="cna-profile-visualization">CNA profile visualization<a class="anchor" aria-label="anchor" href="#cna-profile-visualization"></a>
</h3>
<p>Per-cluster CNA profiles provide a detailed view of copy number
alterations within each subpopulation:</p>
<ul>
<li>Genome-wide coverage, each dot represents a feature, showing log R
ratio values</li>
<li>Allele-specific imbalance, each dot represents a heterozygous
variant, highlighting allelic deviations</li>
<li>Chromosome separation</li>
<li>Segmentation of the genome based on the coverage and allele
signals</li>
<li>Estimated total and minor copy numbers inferred for each
segment</li>
</ul>
<p>In addition to per-cluster profiles, plotting all cells together can
provide a global overview but should not replace cluster-level
interpretation.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Per cluster</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">muscadet</span><span class="op">$</span><span class="va">cnacalling</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"figures"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CNAprofile_"</span>, <span class="va">i</span>, <span class="st">".png"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="va">filename</span>, width <span class="op">=</span> <span class="fl">15</span>, height <span class="op">=</span> <span class="fl">7.5</span>, units <span class="op">=</span> <span class="st">"in"</span>, res <span class="op">=</span> <span class="fl">150</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="../reference/plotProfile.html">plotProfile</a></span><span class="op">(</span><span class="va">muscadet</span>, data <span class="op">=</span> <span class="va">i</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Cluster:"</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># All cells</span></span>
<span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"figures"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CNAprofile_allcells.png"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="va">filename</span>, width <span class="op">=</span> <span class="fl">15</span>, height <span class="op">=</span> <span class="fl">7.5</span>, units <span class="op">=</span> <span class="st">"in"</span>, res <span class="op">=</span> <span class="fl">150</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotProfile.html">plotProfile</a></span><span class="op">(</span><span class="va">muscadet</span>, data <span class="op">=</span> <span class="st">"allcells"</span>, title <span class="op">=</span> <span class="st">"All cells"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## quartz_off_screen </span></span>
<span><span class="co">##                 2</span></span></code></pre>
<div class="float">
<img src="figures/CNAprofile_1.png" style="width:100.0%" fig-align="center" alt="CNA profile - cluster 1"><div class="figcaption">CNA profile - cluster 1</div>
</div>
<div class="float">
<img src="figures/CNAprofile_2.png" style="width:100.0%" fig-align="center" alt="CNA profile - cluster 2"><div class="figcaption">CNA profile - cluster 2</div>
</div>
<div class="float">
<img src="figures/CNAprofile_3.png" style="width:100.0%" fig-align="center" alt="CNA profile - cluster 3"><div class="figcaption">CNA profile - cluster 3</div>
</div>
<div class="float">
<img src="figures/CNAprofile_allcells.png" style="width:100.0%" fig-align="center" alt="CNA profile - all cells"><div class="figcaption">CNA profile - all cells</div>
</div>
<hr>
<p>Additionnally, a complete CNA profile of all clusters at the
sample-level can be visualize using <code><a href="../reference/plotCNA.html">plotCNA()</a></code>.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CNAplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCNA.html">plotCNA</a></span><span class="op">(</span><span class="va">muscadet</span>, cf.gradient <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"figures/CNAplot.png"</span>, <span class="va">CNAplot</span>, width <span class="op">=</span> <span class="fl">7</span>, height <span class="op">=</span> <span class="fl">4</span>, units <span class="op">=</span> <span class="st">"in"</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="figures/CNAplot.png" style="width:100.0%" fig-align="center" alt="Complete CNA profile"><div class="figcaption">Complete CNA profile</div>
</div>
</div>
<div class="section level3">
<h3 id="integrating-cna-clusters-with-atacrna-embeddings">Integrating CNA clusters with ATAC/RNA embeddings<a class="anchor" aria-label="anchor" href="#integrating-cna-clusters-with-atacrna-embeddings"></a>
</h3>
<p>After CNA calling, CNA-defined clusters can be projected onto
independent ATAC- and RNA-based UMAP embeddings. This visualization
helps to explore the relationship between genomic alterations and
cellular phenotypes.</p>
<p>This step illustrates:</p>
<ul>
<li>How CNA-defined subclones relate to transcriptional and chromatin
states</li>
<li>Whether CNA structure aligns with modality-specific embeddings</li>
<li>Potential decoupling between copy-number variation and
expression</li>
</ul>
<p>This projection provides a complementary perspective to cluster-level
CNA analysis, integrating genomic and functional information in a single
visualization.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># import UMAP coordinates based on chromatin accessibility and expression</span></span>
<span><span class="va">UMAP_coord_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"UMAP_coord_atac.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">UMAP_coord_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">inputdir</span>, <span class="st">"UMAP_coord_rna.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add cluster assignments to coordinates tables</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="va">muscadet</span><span class="op">$</span><span class="va">cnacalling</span><span class="op">$</span><span class="va">clusters</span></span>
<span><span class="va">UMAP_coord_atac</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">UMAP_coord_atac</span><span class="op">$</span><span class="va">cell</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">UMAP_coord_rna</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">UMAP_coord_rna</span><span class="op">$</span><span class="va">cell</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">palette_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#FABC2A"</span>, <span class="st">"#7FC97F"</span>, <span class="st">"#EE6C4D"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">UMAP_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">UMAP_coord_atac</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">UMAP_1</span> , y <span class="op">=</span> <span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">""</span>, values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">palette_clusters</span>, <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span></span>
<span>    <span class="va">UMAP_coord_atac</span><span class="op">$</span><span class="va">cluster</span></span>
<span>    <span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"scATAC-seq"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"UMAP 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"UMAP 2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    <span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"figures/UMAP_atac.png"</span>, <span class="va">UMAP_atac</span><span class="op">)</span></span>
<span></span>
<span><span class="va">UMAP_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">UMAP_coord_rna</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">UMAP_1</span> , y <span class="op">=</span> <span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">""</span>, values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">palette_clusters</span>, <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span></span>
<span>        <span class="va">UMAP_coord_rna</span><span class="op">$</span><span class="va">cluster</span></span>
<span>    <span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"scRNA-seq"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"UMAP 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"UMAP 2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"figures/UMAP_rna.png"</span>, <span class="va">UMAP_rna</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/UMAP_atac.png" style="width:60.0%" fig-align="center" alt="Copy number clusters projected on ATAC UMAP"><img src="figures/UMAP_rna.png" style="width:60.0%" fig-align="center" alt="Copy number clusters projected on RNA UMAP"></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="key-parameters-to-tune">Key parameters to tune<a class="anchor" aria-label="anchor" href="#key-parameters-to-tune"></a>
</h2>
<table class="table">
<colgroup>
<col width="9%">
<col width="29%">
<col width="25%">
<col width="35%">
</colgroup>
<thead><tr class="header">
<th><strong>Step</strong></th>
<th><strong>Parameter</strong></th>
<th><strong>Argument</strong></th>
<th><strong>Notes</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>Compute log R ratio</strong></td>
<td>Feature filters: remove low-quality or poorly covered features</td>
<td>
<code>minReads</code>, <code>minPeaks</code> (ATAC);
<code>refReads</code>, <code>refMeanReads</code> (RNA)</td>
<td>Adjust depending on dataset coverage</td>
</tr>
<tr class="even">
<td></td>
<td>Window/smoothing: define size of genomic windows</td>
<td>
<code>windowSize</code>, <code>slidingSize</code> (ATAC);
<code>genesPerWindow</code> (RNA)</td>
<td>Choose based on desired resolution of CNAs</td>
</tr>
<tr class="odd">
<td><strong>Clustering</strong></td>
<td>Range for partitions: explore multiple resolutions or cluster
numbers</td>
<td>
<code>res_range</code> (Seurat-based); <code>k_range</code>
(hclust)</td>
<td>Select based on cluster stability and biological
interpretability</td>
</tr>
<tr class="even">
<td></td>
<td>Selected dimensions from low-dimensional embeddings (Seurat-based
only)</td>
<td>
<code>dims_list</code> (Seurat-based)</td>
<td>Choose top dimensions capturing most variance; adjust for noise and
dataset complexity</td>
</tr>
<tr class="odd">
<td><strong>CNA calling</strong></td>
<td>Omic selection for coverage: use the most reliable omic to reduce
noise</td>
<td><code>omics.coverage</code></td>
<td>
<code>"ATAC"</code> or <code>"RNA"</code>; multiple omics can be
used if both are reliable</td>
</tr>
<tr class="even">
<td></td>
<td>Feature/position filters: remove poorly covered positions or
variants</td>
<td><code>depthmin.*.*</code></td>
<td>Adjust depending on sequencing depth and data quality</td>
</tr>
</tbody>
</table>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Marie Denoulet, Nils Giordano, Stéphane Minvielle, Céline Vallot, Eric Letouzé.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
