#' Heatmap plot for `muscadet` object
#'
#' This function generates a heatmap to visualize log R ratio (LRR) data
#' contained in \code{\link{muscadet}} objects. One heatmap is generated per
#' omic, rows are cells and columns are chromosomes, for `muscadet` object
#' containing multiple omics, the heatmaps are plotted horizontally aligned. The
#' cells can be clustered for a specific k number of cluster following the
#' clustering step of `muscadet` object, or custom cluster assignments can be
#' used. Additionally, LRR values from bulk sequencing data can be plotted as an
#' annotation under the heatmaps.
#'
#' @param x A \code{\link{muscadet}} object containing LRR data for all omics
#'   (with [muscadet::computeLogRatio()]) and clustering data (with
#'   [muscadet::clusterMuscadet()]) (`muscadet`).
#'
#' @param filename (Optional) Character string specifying the file path to save
#'   the heatmap image in the PNG format (`character` string).
#'
#' @param k (Optional) Integer specifying the cluster number to plot
#'   (`integer`). It should be within the range of k used for clustering (with
#'   [muscadet::clusterMuscadet()]). Should be provided if `clusters` is `NULL`.
#'
#' @param clusters (Optional) A custom named vector of cluster assignments
#'   (`integer` named vector). Names must corresponds the cell names of the
#'   `muscadet` object: to the totality of cells if `show_missing = TRUE`, or to
#'   only common cells (cells with data in all omics) if `show_missing = FALSE`.
#'   Should be provided if `k` is `NULL`.
#'
#' @param title Character string for the title of the heatmap (`character`
#'   string). Default is an empty character string.
#'
#' @param add_bulk_lrr Logical value indicating whether to add bulk LRR data as
#'   annotations if present in the muscadet object (`logical`). By default, it
#'   is set to `NULL`, which will automatically include bulk LRR data if it is
#'   available in the muscadet object.
#'
#' @param show_missing Logical value indicating whether to show missing cells
#'   (cells with missing data in at least one omic) in the heatmaps (`logical`).
#'   Default is `TRUE`.
#'
#' @param white_scale Numeric vector of length 2 or a list of numeric vectors
#'   (`numeric` vector or `list`).
#' - If a numeric vector of length 2, the same white color boundaries are
#'   applied to all omics in the muscadet object. E.g. `c(0.3, 0.7)` (default).
#' - If a list (named or not with omics name), it must have the same length as
#'   the number of omics in the muscadet object, where each vector element
#'   applies the white color boundaries for a specific omic. E.g. `list(c(0.3,
#'   0.7), c(0.4, 0.6))` uses 0.3 and 0.7 quantiles of LRR ref data for the 1st
#'   omic heatmap, and 0.4 and 0.6 quantiles for the second.
#'
#'   Values of the vectors must be between 0 and 1, specifying the quantiles of
#'   the LRR reference data that define the boundaries for the white color in
#'   each heatmap. LRR values falling within this range are considered close to
#'   the majority of the LRR reference data, indicating no significant gain or
#'   loss of coverage, and are represented as white on the heatmap. Default is
#'   `c(0.3, 0.7)`.
#'
#'
#' @param colors Vector of colors for the cluster annotation (`character`
#'   vector). Default is `NULL`, which uses predefined colors.
#'
#' @param quiet Logical value indicating whether to suppress messages. Default
#'   is `FALSE`.
#'
#' @return A list containing:
#' - `plot`: A \code{\link{gTree}} object created with [grid::grid.grab()] (\code{\link{gTree}}).
#' - `width`: Width of the heatmap plot in mm (\code{\link{unit}}).
#' - `height`: Height of the heatmap plot in mm (\code{\link{unit}}).
#'
#' If the `filename` argument is provided, the heatmap is directly saved as a
#' PNG image at the provided path.
#'
#' @import ComplexHeatmap
#' @importFrom circlize colorRamp2
#' @importFrom methods slot
#' @importFrom stats median
#' @importFrom grDevices pdf png palette dev.off
#' @importFrom grid gpar unit grid.rect grid.text grid.grab
#'
#' @export
#'
#' @examples
#' # Load example muscadet object
#' data(muscadet_obj)
#'
#' # Perform clustering (if not already done)
#' muscadet_obj <- clusterMuscadet(muscadet_obj, k_range = 2:5)
#'
#' # Custom title with clustering parameters
#' title <- paste("Example sample |",
#'      slot(muscadet_obj, "clustering")[["params"]][["dist_method"]],
#'      slot(muscadet_obj, "clustering")[["params"]][["hclust_method"]], "|",
#'      "k=3", "|",
#'      "Weights of omics:",
#'      paste(slot(muscadet_obj, "clustering")[["params"]][["weights"]], collapse = ", "))
#'
#' # Generate heatmap
#' heatmapMuscadet(
#'     muscadet_obj,
#'     k = 3,
#'     filename = file.path("heatmap_muscadet_k3.png"),
#'     title = title
#' )
#'
#' heatmapMuscadet(
#'     muscadet_obj,
#'     k = 3,
#'     filename = file.path("heatmap_muscadet_k3_commoncells.png"),
#'     title = title,
#'     show_missing = FALSE
#' )
#'
#' # Loop over k
#' for (k in names(muscadet_obj$clustering$clusters)) {
#'     filename <- paste0("heatmap_muscadet_k", k, ".png")
#'     title <- paste(
#'         "Example sample |",
#'         muscadet_obj$clustering$params[["dist_method"]],
#'         muscadet_obj$clustering$params[["hclust_method"]],
#'         "|",
#'         paste0("k=", k),
#'         "|",
#'         "Weights of omics:",
#'         paste(muscadet_obj$clustering$params[["weights"]], collapse = ", ")
#'     )
#'     heatmapMuscadet(
#'         muscadet_obj,
#'         k = as.integer(k),
#'         filename = filename,
#'         title = title
#'     )
#' }
#'
heatmapMuscadet <- function(x,
                            filename = NULL,
                            k = NULL,
                            clusters = NULL,
                            title = "",
                            add_bulk_lrr = NULL,
                            show_missing = TRUE,
                            white_scale = c(0.3, 0.7),
                            colors = NULL,
                            quiet = FALSE) {
    # Validate input: x must be a muscadet object
    stopifnot("Input object `x` must be of class `muscadet`." = inherits(x, "muscadet"))

    # Validate the muscadet object contains clustering results
    stopifnot(
        "The muscadet object `x` does not contain clustering data (use `clusterMuscadet()` to perform clustering of log R ratio data)." =
            !is.null(slot(x, "clustering"))
    )

    # Validate the clustering result for the specified k
    stopifnot(
        "The muscadet object `x` must contain clustering results for the specified `k`." =
            as.character(k) %in% names(slot(x, "clustering")[["clusters"]])
    )

    # Set to no missing cells if only one omic
    if (length(x@omics) == 1)
        show_missing <- FALSE

    # Validate k and clusters
    stopifnot("Both `k` and `clusters` cannot be NULL." = !(is.null(k) &&
                                                                is.null(clusters)))

    # Validate k in clustering slot
    if (!is.null(k)) {
        stopifnot(
            "The muscadet object `x` must contain the clustering results for the `k` provided." = as.character(k) %in% names(slot(x, "clustering")[["clusters"]])
        )
    }

    # Default addition of bulk data
    if (is.null(add_bulk_lrr)) {
        add_bulk_lrr <- !is.null(x@bulk.data$log.ratio)
    }

    # Validate white_scale
    if (is.numeric(white_scale) && length(white_scale) == 2) {
        stopifnot("`white_scale` must be a numeric vector of length 2." = length(white_scale) == 2)
        stopifnot(
            "`white_scale` values must be between 0 and 1." = all(white_scale >= 0 &
                                                                      white_scale <= 1)
        )
        # Single pair of values applied to all omics
        white_scale <- round(sort(white_scale), 2)
        white_scale <- rep(list(white_scale), length(x@omics))  # Ensure list matches omics count
        names(white_scale) <- names(x@omics)  # Assign omic names for consistency

    } else if (is.list(white_scale)) {
        stopifnot(
            "`white_scale` must have the same length as the number of omics in muscadet object `x`." =
                length(white_scale) == length(x@omics)
        )

        stopifnot("All elements of `white_scale` must be numeric vectors of length 2." =
                      all(
                          vapply(white_scale, function(x)
                              is.numeric(x) && length(x) == 2, logical(1))
                      ))

        white_scale <- lapply(white_scale, function(x) {
            x <- round(x, 2)
            stopifnot("The two elements of `white_scale` vectors must not be equal." = x[1] != x[2])
            if (x[1] > x[2])
                x <- sort(x)
            return(x)
        })

        # If white_scale is unnamed, assign names based on omics order
        if (is.null(names(white_scale))) {
            names(white_scale) <- names(x@omics)
        } else {
            stopifnot(
                "Named `white_scale` list must have names matching the omics in muscadet object `x`." =
                    all(names(white_scale) %in% names(x@omics))
            )
        }
    } else {
        stop(
            "`white_scale` must be either a numeric vector of length 2 or a list of such vectors."
        )
    }


    # Set default color palette for clusters if not provided
    if (is.null(colors)) {
        colors <- c(
            "#FABC2A",
            "#7FC97F",
            "#EE6C4D",
            "#39ADBD",
            "#BEAED4",
            "#FEE672",
            "#F76F8E",
            "#487BEA",
            "#B67BE6",
            "#F38D68",
            "#7FD8BE",
            "#F2AFC0"
        )
    }
    # Set palette
    palette(colors)

    # Check if output directory exists
    if (!is.null(filename))
        stopifnot("The directory doesn't exist" = file.exists(dirname(filename)))

    if (!grepl(".(png|pdf)$", filename)) {
        stop("The `filename` argument must end with either .png or .pdf.")
    }

    # Get common and all cells in clustering order
    common_cells <- sort(Reduce(intersect, lapply(muscadet::matLogRatio(x), colnames)))
    all_cells <- sort(Reduce(union, lapply(muscadet::matLogRatio(x), colnames)))

    if (quiet == FALSE) {
        # Print information messages
        message("---- Heatmap Parameters ----")
        message("Omics in the muscadet object: ",
                paste(sapply(slot(x, "omics"), function(n)
                    slot(n, "label.omic")), collapse = ", "))
        omic_dims <- sapply(slot(x, "omics"), function(m) {
            paste0(
                m@label.omic,
                ": ",
                ncol(muscadet::matLogRatio(m)),
                " cells x ",
                nrow(muscadet::matLogRatio(m)),
                " features"
            )
        })
        message("Omics log R ratio data dimensions:\n  ",
                paste(omic_dims, collapse = "\n  "))

        if (!is.null(k)) message("Number of clusters (k): ", k)
        if (!is.null(clusters)) message("Custom clusters provided: ", length(unique(clusters)), " clusters.")

        message(
            "Number of cells: ",
            length(all_cells),
            " total (",
            length(common_cells),
            " common across all omics)."
        )

        message("Show missing cells: ", show_missing)
        message("Bulk LRR annotations: ",
                ifelse(add_bulk_lrr, slot(x, "bulk.data")[["label"]], add_bulk_lrr))
        message("White scale quantiles: ", paste(white_scale, collapse = " - "))
        message("Output file: ", ifelse(is.null(filename), "None", filename))
    }

    # Create list of heatmap objects
    list_ht <- lapply(names(x@omics), function(omic_name) {
        muscomic <- x@omics[[omic_name]]

        pdf(file = NULL)
        ht_opt(message = F)

        # Get chromosome info for features
        coord <- muscomic@coverage$coord.features
        chrom <- factor(coord[coord$keep, "CHROM"], levels = unique(coord[coord$keep, "CHROM"]))

        # Define color breaks for heatmap using white_scale argument
        col_breaks <- c(-5, muscomic@coverage$ref.log.ratio.perc[as.character(white_scale[[omic_name]])], 5)

        # Extract log R ratio matrix
        if (show_missing == TRUE) {
            mat <- t(muscadet::matLogRatio(muscomic))
            cells.diff <- setdiff(all_cells, rownames(mat)) # identify missing cells
            if (length(cells.diff) > 0) {
                mat.na <- matrix(
                    data = NA,
                    nrow = length(cells.diff),
                    ncol = ncol(mat),
                    dimnames = list(cells.diff, colnames(mat))
                )
                mat <- rbind(mat, mat.na)[all_cells, ] # add empty rows to matrix
            } else {
                mat <- mat[all_cells, ]
            }
        }
        if (show_missing == FALSE) {
            mat <- t(muscadet::matLogRatio(muscomic))
            mat <- mat[common_cells, ]
        }

        # Create heatmap
        ht <- ComplexHeatmap::Heatmap(
            mat,
            name = muscomic@label.omic,
            heatmap_legend_param = list(title = muscomic@label.omic),
            show_column_names = F,
            show_row_names = F,
            cluster_columns = F,
            column_split = chrom,
            column_title = paste(
                muscomic@label.omic,
                "coverage on",
                muscomic@coverage$label.features
            ),
            row_title_gp = gpar(fontsize = 10),
            column_title_gp = gpar(fontsize = 10),
            border_gp = gpar(col = "black", lwd = 1),
            heatmap_height = unit(12, "cm"),
            heatmap_width = unit(18, "cm"),
            col = circlize::colorRamp2(col_breaks, c(
                "#00008E", "white", "white", "#630000"
            )),
            row_title_rot = 0,
            raster_device = "png",
            raster_quality = 3
        )

        # Add empty chromosome annotation
        emp <- ComplexHeatmap::anno_empty(border = FALSE, height = unit(2, "mm"))
        ht <- ComplexHeatmap::attach_annotation(ht,
                                                columnAnnotation(
                                                    emp_annot = emp,
                                                    name = paste0("chrom_", muscomic@label.omic)
                                                ),
                                                side = "top")
        # Rename chromosome annotation
        ht@top_annotation@anno_list[["emp_annot"]]@name <- paste0("chrom_", muscomic@label.omic)
        ht@top_annotation@anno_list[["emp_annot"]]@label <- paste0("chrom_", muscomic@label.omic)
        ht@top_annotation@anno_list[["emp_annot"]]@name_param[["label"]] <- paste0("chrom_", muscomic@label.omic)
        names(ht@top_annotation@anno_list)[1] <- paste0("chrom_", muscomic@label.omic)

        # Add bulk LRR data as annotation
        if (add_bulk_lrr == TRUE) {
            # Retrieve bulk lrr values on features
            bulk_df <- muscadet::getLogRatioBulk(muscomic, x@bulk.data$log.ratio)
            # Define color scale
            bulk_col <- list(circlize::colorRamp2(
                c(
                    min(bulk_df$bulk.lrr, na.rm = TRUE),
                    median(bulk_df$bulk.lrr, na.rm = TRUE),
                    max(bulk_df$bulk.lrr, na.rm = TRUE)
                ),
                c("#00008E", "white", "#630000")
            ))
            names(bulk_col) = paste0("bulk_", muscomic@label.omic)
            # Create data frame for annotation
            bulk_df <- as.data.frame(bulk_df$bulk.lrr)
            colnames(bulk_df) <- paste0("bulk_", muscomic@label.omic)
            # Create bulk LRR annotation
            annot_bulk <- ComplexHeatmap::HeatmapAnnotation(
                df = bulk_df,
                col = bulk_col,
                annotation_label = x@bulk.data$label,
                annotation_name_side = "left",
                annotation_name_gp = gpar(cex = 0.7)
            )
            # Attach annotation to heatmap
            ht <- ComplexHeatmap::attach_annotation(ht, annot_bulk, side = "bottom")
        }

        dev.off()
        return(ht)
    })


    # Create empty annotation to add number of cells (ncells)
    ha <- ComplexHeatmap::rowAnnotation(ncells = anno_empty(border = FALSE, width = unit(12, "mm")))
    # Combine heatmaps and ncells empty annotation
    ht_list <- ha + Reduce("+", list_ht)

    # Create layout for the list of heatmaps
    pdf(file = NULL)
    if (show_missing == TRUE) {
        if (is.null(clusters)) {
            # 1. Cluster assignments from the muscadet object clustering with all cells
            clusters <- x@clustering$clusters[[as.character(k)]]
        }
        n_cells <- table(clusters)

        ht_all <- ComplexHeatmap::draw(
            ht_list,
            column_title = title,
            ht_gap = unit(1, "cm"),
            row_split = factor(clusters[all_cells], levels =
                                   sort(unique(clusters))),
            row_order = names(clusters),
            cluster_rows = F,
            merge_legend = TRUE
        )


    } else if (show_missing == FALSE) {
        if (is.null(clusters)) {
            # 2. Clustering hclust from the muscadet object clustering with common cells
            hc <- x@clustering$hclust # hclust object to print the dendrogram on the heatmap
            n_cells <- table(dendextend::cutree(hc, k, order_clusters_as_data = FALSE))

            ht_all <- ComplexHeatmap::draw(
                ht_list,
                column_title = title,
                ht_gap = unit(1, "cm"),
                cluster_rows = hc,
                row_split = as.integer(k),
                row_dend_reorder = FALSE,
                merge_legend = TRUE
            )
        } else {
            # 3. Custom cluster assignments vector
            n_cells <- table(clusters)
            ht_all <- ComplexHeatmap::draw(
                ht_list,
                column_title = title,
                ht_gap = unit(1, "cm"),
                row_split = factor(clusters[common_cells], levels =
                                       sort(unique(
                                           clusters
                                       ))),
                row_order = names(clusters),
                cluster_rows = F,
                merge_legend = TRUE
            )
        }
    }
    dev.off()

    # Save complete plot of heatmaps as PNG
    if (!is.null(filename)) {
        png(
            filename = filename,
            width = ht_all@ht_list_param[["width"]],
            height = ht_all@ht_list_param[["height"]],
            units = "mm",
            res = 300
        )
    } else {
        pdf(file = NULL)
    }

    # Print plot
    print(ht_all)

    # Add annotation: number of cells per cluster
    for (i in 1:length(n_cells)) {
        ComplexHeatmap::decorate_annotation("ncells", slice = i, envir = environment(), {
            grid.rect(
                x = 1,
                width = unit(2, "mm"),
                gp = gpar(fill = colors[i], col = NA),
                just = "right"
            )
            grid.text(
                n_cells[i],
                x = 0.7,
                just = "right",
                gp = gpar(cex = 0.75)
            )
        })
    }

    # Add annotation: chromosome numbers
    for (ht_name in names(ht_all@ht_list)[-1]) {
        chrom_names <- unique(names(ht_all@ht_list[[ht_name]]@column_order_list))
        for (i in 1:length(chrom_names)) {
            ComplexHeatmap::decorate_annotation(paste0("chrom_", ht_name),
                                                slice = i,
                                                envir = environment(),
                                                {
                                                    grid.text(chrom_names[i],
                                                              just = "center",
                                                              gp = gpar(fontsize = 8.5))
                                                })
        }
    }

    # store plot object
    if (is.null(filename)) {
        plot.obj <- list(
            plot = grid.grab(),
            width = ht_all@ht_list_param[["width"]],
            height = ht_all@ht_list_param[["height"]]
        )
    }

    dev.off()

    # return the plot object
    if (is.null(filename)) {
        return(plot.obj)
    }
}





#' Silhouette plot for `muscadet` object
#'
#' Generate a silhouette plot for a specified clustering result within a
#' `muscadet` object.
#'
#' @param x A \code{\link{muscadet}} object containing clustering data (using
#'   [muscadet::clusterMuscadet()]).
#'
#' @param k Integer specifying the number of clusters to plot (`integer`).
#'   It should be within the range of k used for clustering (with
#'   [muscadet::clusterMuscadet()]).
#'
#' @param colors Vector of colors for the cluster annotation (`character`
#'   vector). Default is `NULL`, which uses predefined colors.
#'
#' @param title Character string for the title of the plot (`character`
#'   string). If `NULL`, a default title is generated.
#'
#' @return A ggplot object representing the silhouette plot.
#'
#' @import ggplot2
#' @importFrom stats aggregate
#' @importFrom rlang .data
#' @export
#'
#' @examples
#' library("ggplot2")
#'
#' # Load example muscadet object
#' data(muscadet_obj)
#' plotSil(muscadet_obj, k = 3)
#'
#' # Loop over k
#' for (k in names(muscadet_obj$clustering$clusters)) {
#'     plot <- plotSil(muscadet_obj, k = as.integer(k))
#'     ggsave(paste0("plot_silhouette_", k, ".png"), plot)
#' }
#'
plotSil <- function(x, k, colors = NULL, title = NULL) {
    # Validate input: x must be a muscadet object
    stopifnot("Input object `x` must be of class `muscadet`." = inherits(x, "muscadet"))

    # Validate the muscadet object contains clustering results
    stopifnot(
        "The muscadet object `x` does not contain clustering data (use `clusterMuscadet()` to perform clustering of log R ratio data)." =
            !is.null(slot(x, "clustering"))
    )

    # Validate the clustering result for the specified k
    stopifnot(
        "The muscadet object `x` must contain clustering results for the specified `k`." =
            as.character(k) %in% names(slot(x, "clustering")[["clusters"]])
    )

    # Set default color palette for clusters if not provided
    if (is.null(colors)) {
        colors <- c(
            "#FABC2A",
            "#7FC97F",
            "#EE6C4D",
            "#39ADBD",
            "#BEAED4",
            "#FEE672",
            "#F76F8E",
            "#487BEA",
            "#B67BE6",
            "#F38D68",
            "#7FD8BE",
            "#F2AFC0"
        )
    }

    # Generate a default title if none is provided
    if (is.null(title)) {
        title <- paste("Silhouette Plot (k =", as.character(k), ")")
    }

    # Extract silhouette data for the specified k
    sil <- x@clustering[["silhouette"]][["sil.obj"]][[as.character(k)]]
    df <- as.data.frame(sil[, 1:3], stringsAsFactors = TRUE)

    # Order data for plotting
    df <- df[order(df$cluster, -df$sil_width), ]
    df$name <- factor(rownames(df), levels = rownames(df))
    df$cluster <- as.factor(df$cluster)

    # Calculate average silhouette width per cluster
    avg_sil_width <- stats::aggregate(df$sil_width ~ df$cluster, data = df, mean)
    colnames(avg_sil_width) <- c("cluster", "sil_width")

    # Determine positions for cluster annotations
    cluster_counts <- table(df$cluster)
    cluster_positions <- sapply(unique(df$cluster), function(i) {
        # Reverse y-axis positions since it is flipped
        rev_y_position <- sum(cluster_counts) -
            (sum(cluster_counts[as.numeric(levels(df$cluster)[1:i])]) - cluster_counts[i] / 2)
        rev_y_position
    })

    # Create the ggplot object
    p <- ggplot(df,
                aes(
                    x = .data$sil_width,
                    y = .data$name,
                    fill = .data$cluster
                )) +
        geom_bar(stat = "identity", width = 1) +
        scale_y_discrete(limits = rev) +
        theme_bw() +
        theme(
            axis.text.y = element_blank(),
            axis.ticks.y = element_blank(),
            panel.border = element_blank(),
            panel.grid = element_blank(),
            plot.title = element_text(hjust = 0.5),
            plot.subtitle = element_text(hjust = 0.5),
            plot.margin = unit(c(0.5, 1, 0.5, 0.5), "cm"),
            legend.position = "none"
        ) +
        scale_fill_manual(values = colors) +
        labs(
            x = "Silhouette Width",
            y = "",
            title = title,
            subtitle = paste0(
                "Average Silhouette Width = ",
                round(mean(df$sil_width), 4),
                "\n",
                "Annotation: cluster | number of cells | average silhouette width"
            )
        ) +
        ggplot2::xlim(c(min(df$sil_width), max(df$sil_width) + 0.25)) +
        geom_vline(
            xintercept = mean(df$sil_width),
            linetype = "dashed",
            color = "red"
        )

    # Add annotations for each cluster
    for (i in seq_along(cluster_positions)) {
        cluster <- unique(df$cluster)[i]
        avg_width <- round(avg_sil_width$sil_width[avg_sil_width$cluster == cluster], 4)
        count <- cluster_counts[i]
        p <- p + geom_text(
            label = paste(cluster, "|", count, "|", avg_width),
            x = max(df$sil_width) + 0.05,
            y = cluster_positions[i],
            inherit.aes = FALSE,
            hjust = 0,
            vjust = -0.5,
            check_overlap = TRUE
        )
    }

    return(p)
}



#' Plot clustering validation indexes for a `muscadet` object
#'
#' It generates a plot of clustering validation indexes for a clustering result
#' within a `muscadet` object. The index values are computed only using
#' distances between common cells across omics in the `muscadet` object.
#'
#' @param x A \code{\link{muscadet}} object containing clustering data
#'   (generated using [muscadet::clusterMuscadet()]).
#'
#' @param index Character vector specifying one or more validation indexes to
#'   plot among "silhouette", "dunn2", "davisbouldin", "pearsongamma", and "c".
#'   If `NULL`, by default all available indexes are included. If multiple
#'   indexes are selected, the values are normalized for comparability.
#'
#' @param colors Vector of colors for each index in the plot (`character`
#'   vector). Default is `NULL`, which uses predefined colors for the indexes.
#'
#' @param title Character string for the title of the plot (`character`
#'   string). If `NULL`, a default title is generated.
#'
#' @return A ggplot object visualizing the clustering validation indexes across
#'   different numbers of clusters (`k`).
#'
#' @details The function computes several clustering validation indexes,
#'   including:
#'   \itemize{
#'     \item \strong{Silhouette}: Measures how similar an object is to its own
#'     cluster compared to others (see [cluster::silhouette]).
#'     \item \strong{Dunn2}: The ratio of the smallest distance between
#'     observations in different clusters to the largest within-cluster distance
#'     (see [fpc::cluster.stats] `$dunn2`).
#'     \item \strong{Davis-Bouldin}: Measures cluster compactness and separation
#'     (see [clusterSim::index.DB]).
#'     \item \strong{Pearson's Gamma}: Evaluates the goodness of clustering
#'     based on correlation (see [fpc::cluster.stats] `$pearsongamma`).
#'     \item \strong{C Index}: Measures the clustering quality compared to
#'     random data (see [clusterSim::index.C]).
#'   }
#'
#'   If multiple indexes are selected, the values are normalized to fall between
#'   0 and 1. For indexes that are better when minimized ("pearsongamma" and
#'   "c"), their values are reversed for easier comparison.
#'   The k for which the mean of indexes is maximal is highlighted with a dot.
#'
#' @import ggplot2
#' @importFrom stats as.dist
#' @importFrom cluster silhouette
#' @importFrom fpc cluster.stats
#' @importFrom clusterSim index.DB index.C
#' @importFrom tidyr pivot_longer
#'
#' @export
#'
#' @examples
#' # Load example muscadet object
#' data(muscadet_obj)
#'
#' # Plot all indexes
#' plotIndexes(muscadet_obj)
#'
#' # Plot specific indexes
#' plotIndexes(muscadet_obj, index = "silhouette")
#'
plotIndexes <- function(x,
                        index = NULL,
                        colors = NULL,
                        title = NULL) {
    # Validate input: x must be a muscadet object
    stopifnot("Input object `x` must be of class `muscadet`." = inherits(x, "muscadet"))

    # Validate the muscadet object contains clustering results
    stopifnot(
        "The muscadet object `x` does not contain clustering data (use `clusterMuscadet()` to perform clustering of log R ratio data)." =
            !is.null(slot(x, "clustering"))
    )

    # Define default indexes if none are provided
    if (is.null(index)) {
        index <- c("silhouette",
                   "dunn2",
                   "davisbouldin",
                   "pearsongamma",
                   "c")
    }
    # Check for indexes correct names
    index <- match.arg(
        index,
        c("silhouette", "dunn2", "davisbouldin", "pearsongamma", "c"),
        several.ok = TRUE
    )

    # Set default colors if not provided
    if (is.null(colors)) {
        colors <- c(
            "silhouette" = "brown",
            "dunn2" = "coral2",
            "davisbouldin" = "tan2",
            "pearsongamma" = "turquoise4",
            "c" = "skyblue4"
        )
    }

    # Generate a default title if none is provided
    if (is.null(title)) {
        title <- "Clustering Validation Indexes Across k"
    }

    # Extract the clustering data
    k_clusters <- as.integer(names(x@clustering$clusters))  # Number of clusters (k)
    dist <- stats::as.dist(x@clustering$dist)  # Distance matrix to dist object

    # Initialize a data frame to store index values
    df_indexes <- data.frame(k = k_clusters)

    # Compute selected indexes for each k
    for (k in k_clusters) {
        # Extract cluster assignments for the current k
        clusters <- x@clustering$clusters[[as.character(k)]]
        clusters <- clusters[names(dist)] # Restrict to common cells in dist

        # Compute each index if selected
        if ("silhouette" %in% index) {
            df_indexes[df_indexes$k == k, "silhouette"] <-
                summary(cluster::silhouette(as.integer(clusters), dist))[["avg.width"]]
        }
        if ("dunn2" %in% index) {
            df_indexes[df_indexes$k == k, "dunn2"] <-
                fpc::cluster.stats(dist, as.integer(clusters))$dunn2
        }
        if ("davisbouldin" %in% index) {
            df_indexes[df_indexes$k == k, "davisbouldin"] <-
                clusterSim::index.DB(dist, as.integer(clusters))$DB
        }
        if ("pearsongamma" %in% index) {
            df_indexes[df_indexes$k == k, "pearsongamma"] <-
                fpc::cluster.stats(dist, as.integer(clusters))$pearsongamma
        }
        if ("c" %in% index) {
            df_indexes[df_indexes$k == k, "c"] <-
                clusterSim::index.C(dist, as.integer(clusters))
        }
    }

    # If multiple indexes selected:
    if (length(index) > 1) {
        # Normalize indexes to values between 0 and 1
        df_indexes[, 2:ncol(df_indexes)] <- apply(df_indexes[, 2:ncol(df_indexes)], 2, function(x) {
            (x - min(x)) / diff(range(x))
        })
        # Reverse indexes that should be minimized, for comparability
        if ("pearsongamma" %in% index) {
            df_indexes <- mutate(df_indexes, pearsongamma = 1 - .data$pearsongamma)
        }
        if ("c" %in% index) {
            df_indexes <- mutate(df_indexes, c = 1 - .data$c)
        }
    }

    # Find optimal k based on selected indexes
    if (ncol(df_indexes) > 2) {
        # Find the k with the maximum mean index value
        opt_k <- df_indexes[which(rowMeans(df_indexes[, 2:ncol(df_indexes)]) == max(rowMeans(df_indexes[, 2:ncol(df_indexes)]))), "k"]
    } else {
        if (colnames(df_indexes)[2] %in% c("pearsongamma", "c")) {
            # Find k with the minimal value for indexes to minimize
            opt_k <- df_indexes[which(df_indexes[, 2] == min(df_indexes[, 2])), "k"]
        } else {
            # Find k with the maximal value for indexes to maximize
            opt_k <- df_indexes[which(df_indexes[, 2] == max(df_indexes[, 2])), "k"]
        }
    }

    # Transform the data frame for plotting
    df_plot <- tidyr::pivot_longer(df_indexes, -k, names_to = "Index", values_to = "Value")
    df_plot$Index <- factor(df_plot$Index, levels = index) # Maintain order of indexes

    # Generate the plot
    plot <- ggplot(df_plot,
                   aes(
                       x = .data$k,
                       y = .data$Value,
                       color = .data$Index,
                       group = .data$Index
                   )) +
        geom_line(linewidth = 1) +
        geom_point(data = df_plot[which(df_plot$k == opt_k), ]) +
        labs(x = "Number of Clusters (k)", y = "Normalized Index Value", title = title) +
        scale_x_continuous(breaks = unique(df_plot$k)) +
        scale_color_manual(
            values = colors,
            labels = c(
                "silhouette" = "Silhouette",
                "dunn2" = "Dunn2",
                "pearsongamma" = "Pearson's Gamma",
                "davisbouldin" = "Davis-Bouldin",
                "c" = "C Index"
            )
        ) +
        theme_classic() +
        theme(legend.position = "right")

    return(plot)
}



#' Plot CNA profiles from muscadet object
#'
#' This function generates a multi-panel plot of copy number alteration (CNA)
#' profiles from a \code{\link{muscadet}} object, including: log R ratios
#' values, log odds ratio (or variant allele frequency), copy numbers and cell
#' fractions.
#'
#' @param x A \code{\link{muscadet}} object containing CNA calling data to be
#'   visualized (generated using [muscadet::cnaCalling()]).
#' @param data Either a cluster identifier to plot data of a cluster or
#'   "allcells" to plot data on all cells.
#' @param title An optional title for the plot. Default is `NULL`.
#' @param allelic.type A character string indicating the allelic metric to plot:
#'   "lor" for log odds ratio or "vaf" for variant allele frequency. Default is
#'   "lor".
#' @param point.size Numeric value specifying the size of points in the plot in
#'   pixel (with pch = "."). Default is `2`.
#' @param chrom.colors A character vector of length 2 defining alternating
#'   chromosome colors. Default is `c("slategrey", "skyblue")`.
#' @param lor.colors A character vector of length 2 for log odds ratio point
#'   colors depending of variant allele frequency in all cells. Use "none" to
#'   use the alternating chromosome colors (defined by `chrom.colors`). Default
#'   is `c("peachpuff2", "paleturquoise3")`.
#' @param cn.colors A character vector of length 2 for total copy number and
#'   minor allele copy number segment colors. Default is `c("black", "brown2")`.
#' @param cf.colors A character vector of length 3 for cellular fraction
#'   gradient (of 10 values): start color of the gradient, end color of the
#'   gradient, and color for normal diploid (depending on the ploidy). Default
#'   is `c("white", "steelblue", "bisque2")`.
#' @param dipLogR.color A character string for the diploid log R ratio line
#'   color. Default is "magenta4".
#' @param seg.color A character string for the color of segment medians. Default
#'   is "brown2".
#'
#' @return A multi-panel plot of CNA profiles is produced.
#'
#' @import graphics
#' @import grDevices
#'
#' @export
#'
#' @examples
#' # Load example muscadet object
#' data(muscadet_obj)
#'
#' # Plot profile for all cells
#' pdf("CNAprofile_allcells.pdf", width = 15, height = 7.5) # Save as PDF
#' plotProfile(muscadet_obj, data = "allcells", title = "Example data - all cells")
#' dev.off()
#'
plotProfile <- function(x,
                        data,
                        title = NULL,
                        allelic.type = "lor",
                        point.size = 2,
                        chrom.colors = c("slategrey", "skyblue"),
                        lor.colors = c("peachpuff2", "paleturquoise3"),
                        cn.colors = c("black", "brown2"),
                        cf.colors = c("white", "steelblue", "bisque2"),
                        dipLogR.color = c("magenta4"),
                        seg.color = c("brown2")) {
    # Argument checks
    stopifnot(
        "Input object `x` must be of class `muscadet`." = inherits(x, "muscadet"),
        "Invalid `allelic.type'. Use allelic.type = \"lor\" or allelic.type = \"vaf\"." = allelic.type %in% c("lor", "vaf"),
        "`point.size` must be a numeric value." = is.numeric(point.size) &&
            length(point.size) == 1,
        "`chrom.colors` must be a character vector of length 2." = is.character(chrom.colors) &&
            length(chrom.colors) == 2,
        "`lor.colors` must be a character vector of length 2 or `none`." = (is.character(lor.colors) &&
                                                                                length(lor.colors) == 2) ||
            all(lor.colors == "none"),
        "`cn.colors` must be a character vector of length 2." = is.character(cn.colors) &&
            length(cn.colors) == 2,
        "`cf.colors `must be a character vector of length 3." = is.character(cf.colors) &&
            length(cf.colors) == 3,
        "`dipLogR.color` must be a single character value." = is.character(dipLogR.color) &&
            length(dipLogR.color) == 1,
        "`seg.color` must be a single character value." = is.character(seg.color) &&
            length(seg.color) == 1
    )

    # Extract data -------------------------------------------------------------
    if (data == "allcells") {
        pos <- x@cnacalling$positions.allcells
        segs <- x@cnacalling$segments.allcells
        ploidy <- x@cnacalling$ploidy.allcells
        dipLogR <- x@cnacalling$dipLogR.allcells
    } else {
        pos <- x@cnacalling$positions
        segs <- x@cnacalling$segments
        ploidy <- x@cnacalling$ploidy.clusters
        dipLogR <- x@cnacalling$dipLogR.clusters
    }
    stopifnot(
        "`data` must be \"allcells\" or a valid cluster identifier." =
            (data == "allcells" || data %in% unique(pos$cluster))
    )
    if (data != "allcells") {
        pos <- pos[which(pos$cluster == data), ]
        segs <- segs[which(segs$cluster == data), ]
    }

    # Adjust chromosomes levels to get only numeric chromosomes
    pos$chrom <- factor(pos$chrom, levels = unique(pos$chrom))
    segs$chrom <- factor(segs$chrom, levels = unique(segs$chrom))
    chromlevels <- levels(pos$chrom)
    levels(pos$chrom) <- 1:length(levels(pos$chrom))
    levels(segs$chrom) <- 1:length(levels(segs$chrom))
    pos$chrom <- as.numeric(pos$chrom)
    segs$chrom <- as.numeric(segs$chrom)

    # Chromosome alternating dual colors
    chrcol <- 1 + rep(segs$chrom - 2 * floor(segs$chrom / 2), segs$num.mark)

    # Chromosomes position boundaries
    chrbdry <- which(diff(pos$chrom) != 0)

    # Segment position boundaries
    segbdry <- cumsum(c(0, segs$num.mark))
    segstart <- segbdry[-length(segbdry)]
    segend <- segbdry[-1]


    # Layout params ------------------------------------------------------------
    def.par <- par(no.readonly = TRUE)
    layout(matrix(rep(1:4, c(9, 9, 6, 1)), ncol = 1))
    par(
        mar = c(0.25, 3, 0.25, 1),
        mgp = c(1.75, 0.6, 0),
        oma = c(3, 1, 1.5, 0)
    )

    # 1- Plot the LRR data -----------------------------------------------------
    plot(
        pos$cnlr,
        pch = ".",
        cex = point.size,
        col = chrom.colors[chrcol],
        ylab = "Log R ratio",
        cex.lab = 1.5,
        xaxt = "n"
    )
    # Add chromosomes boundaries
    abline(v = chrbdry, lwd = 0.25)
    # # Add LRR median
    # abline(h=median(pos$cnlr, na.rm=TRUE), col="green3")
    # Add diploid LRR
    abline(h = dipLogR, col = dipLogR.color)
    # Add LRR segment medians
    segments(
        segstart,
        segs$cnlr.median,
        segend,
        segs$cnlr.median,
        lwd = 1.75,
        col = seg.color
    )

    # 2- Plot the LOR data -----------------------------------------------------
    if (any(lor.colors == "none")) {
        cols <- chrom.colors[chrcol]
    } else {
        cols <- lor.colors[pos$colVAR]
    }
    if (allelic.type == "lor") {
        plot(
            pos$valor,
            pch = ".",
            cex = point.size,
            col = cols,
            ylab = "Log odds ratio",
            cex.lab = 1.5,
            ylim = c(-4, 4),
            xaxt = "n"
        )
        abline(v = chrbdry, lwd = 0.25)
        segments(segstart,
                 sqrt(abs(segs$mafR)),
                 segend,
                 sqrt(abs(segs$mafR)),
                 lwd = 1.75,
                 col = seg.color)
        segments(segstart,-sqrt(abs(segs$mafR)),
                 segend,-sqrt(abs(segs$mafR)),
                 lwd = 1.75,
                 col = seg.color)
    }
    if (allelic.type == "vaf") {
        pos[which(pos$signal == "coverage"), "vafT"] <- NA
        plot(
            pos$vafT,
            pch = ".",
            cex = point.size,
            col = cols,
            ylab = "Variant allele frequency",
            cex.lab = 1.5,
            ylim = c(0, 1),
            xaxt = "n"
        )
        abline(v = chrbdry, lwd = 0.25)
        segments(
            segstart,
            segs$vafT.median,
            segend,
            segs$vafT.median,
            lwd = 1.75,
            col = seg.color
        )
    }

    # 3- Plot the estimated copy numbers and cf --------------------------------

    # Transform to tcn to log scale over 10
    tcn.i <- which(segs$tcn.em > 10 & !is.na(segs$tcn.em))
    if (length(tcn.i) > 0) {
        segs$tcn.em[tcn.i] <- 9 + log10(segs$tcn.em[tcn.i])
    }

    # Transform to lcn to log scale over 5
    lcn.i <- which(segs$lcn.em > 5)
    if (length(lcn.i) > 0) {
        segs$lcn.em[lcn.i] <- 5 + log10(segs$lcn.em[lcn.i])
    }

    plot(
        c(0, nrow(pos)),
        c(0, max(segs$tcn.em, na.rm = T)),
        type = "n",
        ylab = "Copy number",
        cex.lab = 1.5,
        xaxt = "n"
    )
    abline(v = chrbdry, lwd = 0.25)
    # Add lcn
    segments(segstart,
             segs$lcn.em,
             segend,
             segs$lcn.em,
             lwd = 1.75,
             col = cn.colors[2])
    # Add tcn
    segments(segstart,
             segs$tcn.em,
             segend,
             segs$tcn.em,
             lwd = 1.75,
             col = cn.colors[1])
    # Add cf
    plot(
        c(0, nrow(pos)),
        0:1,
        type = "n",
        ylab = "",
        xaxt = "n",
        yaxt = "n"
    )
    mtext(
        "cf",
        side = 2,
        at = 0.5,
        line = 1,
        las = 2,
        cex = 1
    )

    cfpalette <- colorRampPalette(c(cf.colors[1], cf.colors[2]))(10)
    cfcol <- cfpalette[round(10 * segs$cf.em)]
    cfcol[segs$tcn.em == ploidy &
              segs$lcn.em == ploidy / 2] <- cf.colors[3]
    rect(segstart, 0, segend, 1, col = cfcol, border = NA)

    # Add chromosome ticks on x-axis -------------------------------------------

    # Number positions per chromosomes
    nn <- cumsum(table(pos$chrom))
    # Ticks
    axis(
        labels = chromlevels,
        side = 1,
        at = (nn + c(0, nn[-length(nn)])) / 2,
        cex = 1.5
    )
    # Labels
    mtext(side = 1,
          line = 1.75,
          "Chromosomes",
          cex = 1)

    # Add title ----------------------------------------------------------------
    mtext(
        title,
        side = 3,
        line = 0,
        outer = TRUE,
        cex = 1.1
    )

    # Reset layout
    par(def.par)
}




#' Plot CNA segments across clusters from a muscadet object
#'
#' This function visualizes copy number alteration (CNA) segments across
#' clusters based on data stored in a \code{\link{muscadet}} object. It displays
#' CNAs for each clusters and scales the y-axis based on the proportion of cells
#' in each cluster.
#'
#' @param x A \code{\link{muscadet}} object containing CNA calling data to be
#'   visualized (generated using [muscadet::cnaCalling()]).
#' @param title An optional title for the plot. Default is `NULL`.
#' @param cna.colors A vector of 3 colors for CNA states: gain, loss, and cnloh
#'   (or named vector where names are "gain", "loss", and "cnloh" and the values
#'   are their respective colors). Default is `c("gain" = "#EF6F6AFF", "loss" =
#'   "#6699CCFF", "cnloh" = "#44AA99FF")`.
#'
#' @return A ggplot object representing the CNA segments plot.
#'
#' @import ggplot2
#' @importFrom dplyr mutate
#' @importFrom dplyr left_join
#' @importFrom stats complete.cases
#' @importFrom rlang .data
#'
#' @export
#'
#' @examples
#' library("ggplot2")
#'
#' # Load example muscadet object
#' data(muscadet_obj)
#'
#' # Plot CNA segments
#' plot <- plotCNA(muscadet_obj, title = "Copy Number Alterations in Example Data")
#' print(plot)
plotCNA <- function(x,
                    title = NULL,
                    cna.colors = c(
                        "gain" = "#EF6F6AFF",
                        "loss" = "#6699CCFF",
                        "cnloh" = "#44AA99FF"
                    )) {
    # Argument checks
    stopifnot("Input object `x` must be of class `muscadet`." = inherits(x, "muscadet"))

    # Extract genome
    if (x@genome == "hg38") {
        genome_chrom <- hg38_chrom
    }
    if (x@genome == "hg19") {
        genome_chrom <- hg19_chrom
    }
    if (x@genome == "mm10") {
        genome_chrom <- mm10_chrom
    }
    chromSizes <- as.data.frame(genome_chrom)

    # Keep only autosomes to display
    chromSizes$seqnames <- as.character(chromSizes$seqnames)
    chromSizes <- chromSizes[!chromSizes$seqnames %in% c("X", "Y", "M"), ]

    # Chromosomes start coordinates
    chromStarts <- data.frame(
        chrom = factor(chromSizes$seqnames, levels = unique(chromSizes$seqnames)),
        chrom.start = c(0, cumsum(as.numeric(
            chromSizes$width
        )))[-(nrow(chromSizes) + 1)]
    )

    # Ordered chromosomes names
    chromNames <- factor(unique(chromSizes$seqnames),
                         levels = unique(chromSizes$seqnames))

    # Extract data table
    data <- x@cnacalling$table

    # Extract number of cells per cluster
    ncells <- x@cnacalling$ncells

    # Keep only autosomes to display
    data <- data[!data$chrom %in% c("X", "Y", "M"), ]

    # Add chromosomes start and end coordinates on x axis
    df <- dplyr::left_join(data, chromStarts, by = "chrom") %>%
        dplyr::mutate(
            start.x = .data$start + .data$chrom.start,
            end.x = .data$end + .data$chrom.start
        )

    # Remove consensus segs that have no data in a cluster
    df <- df[complete.cases(df$cluster), ]

    # Compute proportions of cells per cluster for y axis
    prop_clus <- unique(df$prop.cluster[!is.na(df$prop.cluster)])
    prop_starts <- c(0, cumsum(prop_clus[-length(prop_clus)]))
    prop_ends <- cumsum(prop_clus)
    names(prop_starts) <- unique(df$cluster[!is.na(df$cluster)])
    names(prop_ends) <- unique(df$cluster[!is.na(df$cluster)])
    props <- unique(c(prop_starts, prop_ends))
    props_breaks <- props[-length(props)] + (diff(props) / 2)

    # Add clusters coordinates on y axis
    df <- df %>%
        dplyr::mutate(start.y = prop_starts[as.character(.data$cluster)], end.y = prop_ends[as.character(.data$cluster)])

    # Ordered levels for CNA states
    df$cna_state <- factor(df$cna_state, levels = c("gain", "loss", "cnloh"))

    # Construct plot
    cna_plot <- ggplot2::ggplot(
        df,
        aes(
            xmin = .data$start.x,
            xmax = .data$end.x,
            ymin = .data$start.y,
            ymax = .data$end.y,
            fill = .data$cna_state,
            alpha = .data$cf.em
        )
    ) +
        geom_rect() +
        # lines between chromosomes
        geom_vline(
            xintercept = chromStarts$chrom.start,
            colour = "grey",
            linewidth = 0.2
        ) +
        # lines between clusters
        geom_hline(
            yintercept = props,
            colour = "black",
            linewidth = 0.2
        ) +
        # chromosomes labels placement in the center of each chromosome
        scale_x_continuous(
            expand = c(0, 0),
            breaks = chromStarts$chrom.start + (chromSizes$width / 2),
            labels = as.character(chromNames)
        ) +
        # method labels in the center
        scale_y_continuous(
            expand = c(0, 0),
            trans = "reverse",
            limits = c(1, 0),
            breaks = props_breaks,
            labels = paste0("cluster ", unique(df$cluster), "\n", ncells, " cells")
        ) +
        # set colors for the calls and remove the name
        {
            if (length(cna.colors) > 0)
                scale_fill_manual(
                    name = "",
                    values = cna.colors,
                    na.value = "white",
                    na.translate = FALSE,
                    drop = FALSE
                )
        } +
        scale_alpha_continuous(
            name = "cell fraction",
            range = c(0, 1),
            limits = c(0, 1),
            breaks = c(0.2, 0.4, 0.6, 0.8, 1)
        ) +
        guides(fill = guide_legend(order = 1), alpha = guide_legend(order = 2)) +
        ## Set axis labels
        labs(title = title) +
        theme_bw() +
        theme(
            panel.grid = element_blank(),
            plot.title = element_text(hjust = 0.5),
            axis.title.y = element_text(size = 10, face = "bold"),
            axis.text.x = element_text(size = 6, face = "bold"),
            axis.text.y = element_text(size = 8, face = "bold"),
            axis.ticks = element_blank(),
            legend.position = "top"
        )

    return(cna_plot)
}


#' Create heatmap and distribution plots of the different steps of computing log
#' R ratios
#'
#' This function generates heatmap and distribution plots for tumor and
#' reference cells for any step of computing log R ratios matrices. The
#' input object corresponds to the output of [computeLogRatioATAC()] or
#' [computeLogRatioRNA()] with the argument `all_steps = TRUE`.
#'
#' @param obj A list provided as output by the [computeLogRatioATAC()] or
#'   [computeLogRatioRNA()] functions with the argument `all_steps = TRUE`. It
#'   includes tumor and reference matrices at each step of the computing of log
#'   R ratio matrices.
#' @param step The step within the `obj` list to use for plotting (`character`
#'   string). It must match one of the names in `obj`.
#' @param filename File path to save the output plot (`character` string). The
#'   file format is inferred from the extension (".png" or ".pdf").
#' @param title Title of the plot (`character` string). If `NULL`, the title is
#'   automatically generated using the provided step argument and its
#'   corresponding value name (`obj[[step]]$name`).
#' @param col_quantiles A numeric vector of length 4, specifying the quantiles
#'   to use for the color breaks in the heatmap (`numeric`). Default is `c(0.1,
#'   0.4, 0.6, 0.9)`.
#' @param col_boundaries A character vector of 4 colors used for the color scale
#'   of the heatmap (`character` vector). Default is `c("#00008E", "white",
#'   "white", "#630000")`.
#'
#' @return The function does not return any value but saves a heatmaps-histograms plot to the specified file.
#'
#' @import ggplot2
#' @importFrom rlang .data
#' @importFrom ComplexHeatmap Heatmap draw
#' @importFrom circlize colorRamp2
#' @importFrom grid unit grid.grab
#' @importFrom patchwork wrap_plots plot_layout plot_annotation
#'
#' @examples
#'
#' # Create muscomic objects
#' atac <- CreateMuscomicObject(
#'     type = "ATAC",
#'     mat_counts = mat_counts_atac_tumor,
#'     allele_counts = allele_counts_atac_tumor,
#'     features = peaks
#' )
#' rna <- CreateMuscomicObject(
#'     type = "RNA",
#'     mat_counts = mat_counts_rna_tumor,
#'     allele_counts = allele_counts_rna_tumor,
#'     features = genes
#' )
#' atac_ref <- CreateMuscomicObject(
#'     type = "ATAC",
#'     mat_counts = mat_counts_atac_ref,
#'     allele_counts = allele_counts_atac_ref,
#'     features = peaks
#' )
#' rna_ref <- CreateMuscomicObject(
#'     type = "RNA",
#'     mat_counts = mat_counts_rna_ref,
#'     allele_counts = allele_counts_rna_ref,
#'     features = genes
#' )
#'
#' # Create muscadet objects
#' muscadet <- CreateMuscadetObject(
#'     omics = list(atac, rna),
#'     bulk.lrr = bulk_lrr,
#'     bulk.label = "WGS",
#'     genome = "hg38"
#' )
#' muscadet_ref <- CreateMuscadetObject(
#'     omics = list(atac_ref, rna_ref),
#'     genome = "hg38"
#' )
#'
#' # Compute log R ratios with `all_steps = TRUE`
#' obj_atac_all <- computeLogRatioATAC(
#'     matTumor = matCounts(muscadet)$ATAC,
#'     matRef = matCounts(muscadet_ref)$ATAC,
#'     peaksCoord = coordFeatures(muscadet)$ATAC,
#'     genome = slot(muscadet, "genome"),
#'     minReads = 1, # low value for example subsampled datasets
#'     minPeaks = 1, # low value for example subsampled datasets
#'     all_steps = TRUE
#' )
#' names(obj_atac_all)
#'
#' # Plot heatmap and distribution of values for Step04
#' heatmapStep(obj = obj_atac_all,
#'             step = "step01",
#'             filename = "step01.png",
#'             title = "Example data - Step 01")
#'
#' @export
#'
heatmapStep <- function(obj,
                        step,
                        filename,
                        title = NULL,
                        col_quantiles = c(0.1, 0.4, 0.6, 0.9),
                        col_boundaries = c("#00008E", "white", "white", "#630000")) {
    # Argument checks
    if (!is.list(obj)) {
        stop("The `obj` argument must be a list.")
    }

    if (!(step %in% names(obj))) {
        stop(paste(
            "The specified `step` (`",
            step,
            "`) is not found in `obj`.",
            sep = ""
        ))
    }

    if (!grepl(".(png|pdf)$", filename)) {
        stop("The `filename` argument must end with either .png or .pdf.")
    }

    if (!is.numeric(col_quantiles) || length(col_quantiles) != 4) {
        stop("The `col_quantiles` argument should be a numeric vector of length 4.")
    }

    if (!is.character(col_boundaries) ||
        length(col_boundaries) != 4) {
        stop("The `col_boundaries` argument should be a character vector of length 4.")
    }

    # Extract the tumor and reference matrices
    matTumor <- t(obj[[step]]$matTumor)
    matRef <- t(obj[[step]]$matRef)
    name <- obj[[step]]$name

    # Generate title if NULL
    if (is.null(title)) {
        title <- paste(step, "-", name)
    }

    # Chromosome factor from coordinates
    coord <- obj$coord
    chrom <- coord[which(coord$id %in% colnames(matTumor)), "CHROM"]
    chrom <- factor(chrom, levels = unique(chrom))

    # Combine matrices and calculate breaks for color scale
    mat <- rbind(matTumor, matRef)
    col_breaks <- quantile(mat, col_quantiles)

    # Color scale function
    col_fun <- circlize::colorRamp2(col_breaks, col_boundaries)

    # Calculate bin_width
    bin_width <- (max(mat) - min(mat)) / 1000

    # Create heatmaps
    ht_Tum <- ComplexHeatmap::Heatmap(
        matTumor,
        name = "Tumor",
        heatmap_legend_param = list(title = name),
        row_title = "Tumor cells",
        row_title_gp = gpar(fontsize = 12),
        show_column_names = FALSE,
        show_row_names = FALSE,
        cluster_columns = FALSE,
        cluster_rows = TRUE,
        column_split = chrom,
        column_title_gp = gpar(fontsize = 10),
        border_gp = gpar(col = "black", lwd = 1),
        heatmap_height = unit(12, "cm"),
        heatmap_width = unit(18, "cm"),
        col = col_fun,
        raster_device = "png",
        raster_quality = 3
    )

    ht_Ref <- ComplexHeatmap::Heatmap(
        matRef,
        name = "Ref",
        heatmap_legend_param = list(title = name),
        row_title = "Reference cells",
        row_title_gp = gpar(fontsize = 12),
        show_column_names = FALSE,
        show_row_names = FALSE,
        cluster_columns = FALSE,
        cluster_rows = TRUE,
        column_split = chrom,
        column_title_gp = gpar(fontsize = 10),
        border_gp = gpar(col = "black", lwd = 1),
        heatmap_height = unit(12, "cm"),
        heatmap_width = unit(18, "cm"),
        col = col_fun,
        raster_device = "png",
        raster_quality = 3
    )

    # Draw heatmaps
    pdf(file = NULL)  # Temporarily create the plot device for heatmap images
    ht_Tum_2 <- ComplexHeatmap::draw(ht_Tum, column_title = paste(name, "in tumor cells"))
    ht_Tum_grob <- grid.grab()  # Capture heatmap as grob
    ht_Ref_2 <- ComplexHeatmap::draw(ht_Ref, column_title = paste(name, "in reference cells"))
    ht_Ref_grob <- grid.grab()  # Capture heatmap as grob
    dev.off()

    # Histogram calculations and data preparation
    data_Tum <- as.data.frame(as.table(matTumor))
    data_Ref <- as.data.frame(as.table(matRef))
    colnames(data_Tum) <- colnames(data_Ref) <- c("Row", "Column", "Value")

    # Create bins based on the bin_width and compute frequency for gradient bar
    data_Tum$Value_bin <- cut(
        data_Tum$Value,
        breaks = seq(floor(min(data_Tum$Value)), ceiling(max(data_Tum$Value)), by = bin_width),
        include.lowest = TRUE,
        right = FALSE,
        labels = FALSE
    )

    gradient_data_Tum <- data.frame(
        x_min = seq(min(matTumor), max(matTumor), length.out = 500)[-500],
        x_max = seq(min(matTumor), max(matTumor), length.out = 500)[-1],
        y_min = rep(-(max(
            table(data_Tum$Value_bin)
        ) * 0.02), 499),
        y_max = rep(-(max(
            table(data_Tum$Value_bin)
        ) * 0.1), 499)
    )
    gradient_data_Tum$fill <- col_fun((gradient_data_Tum$x_min + gradient_data_Tum$x_max) / 2)

    data_Ref$Value_bin <- cut(
        data_Ref$Value,
        breaks = seq(floor(min(data_Ref$Value)), ceiling(max(data_Ref$Value)), by = bin_width),
        include.lowest = TRUE,
        right = FALSE,
        labels = FALSE
    )

    gradient_data_Ref <- data.frame(
        x_min = seq(min(matRef), max(matRef), length.out = 500)[-500],
        x_max = seq(min(matRef), max(matRef), length.out = 500)[-1],
        y_min = rep(-(max(
            table(data_Ref$Value_bin)
        ) * 0.02), 499),
        y_max = rep(-(max(
            table(data_Ref$Value_bin)
        ) * 0.1), 499)
    )
    gradient_data_Ref$fill <- col_fun((gradient_data_Ref$x_min + gradient_data_Ref$x_max) / 2)

    # Create histograms
    hist_Tum <- ggplot(data_Tum, aes(x = .data$Value)) +
        geom_histogram(
            aes(y = after_stat(.data$count)),
            binwidth = bin_width,
            color = "black",
            fill = "black"
        ) +
        geom_rect(
            data = gradient_data_Tum,
            aes(
                xmin = .data$x_min,
                xmax = .data$x_max,
                ymin = .data$y_min,
                ymax = .data$y_max,
                fill = .data$fill
            ),
            inherit.aes = FALSE
        ) +
        scale_fill_identity() +
        labs(
            title = paste(name, "distribution in tumor cells"),
            x = "Value",
            y = "Frequency"
        ) +
        theme_classic()

    hist_Ref <- ggplot(data_Ref, aes(x = .data$Value)) +
        geom_histogram(
            aes(y = after_stat(.data$count)),
            binwidth = bin_width,
            color = "black",
            fill = "black"
        ) +
        geom_rect(
            data = gradient_data_Ref,
            aes(
                xmin = .data$x_min,
                xmax = .data$x_max,
                ymin = .data$y_min,
                ymax = .data$y_max,
                fill = .data$fill
            ),
            inherit.aes = FALSE
        ) +
        scale_fill_identity() +
        labs(
            title = paste(name, "distribution in reference cells"),
            x = "Value",
            y = "Frequency"
        ) +
        theme_classic()

    # Combine heatmap and histogram plots
    final_plot <- patchwork::wrap_plots(c(list(ht_Tum_grob, ht_Ref_grob), list(hist_Tum, hist_Ref)), ncol = 2) +
        plot_layout(nrow = 2, heights = c(3, 1)) +
        plot_annotation(title = title,
                        theme = theme(plot.title = element_text(size = 16)))

    # Output to file
    if (grepl(".png", basename(filename))) {
        png(
            filename = filename,
            width = ht_Tum_2@ht_list_param[["width"]] + ht_Ref_2@ht_list_param[["width"]],
            height = ht_Tum_2@ht_list_param[["height"]] * 1.75,
            units = "mm",
            res = 300
        )
        print(final_plot)
        dev.off()

    } else if (grepl(".pdf", basename(filename))) {
        pdf(
            file = filename,
            width = (ht_Tum_2@ht_list_param[["width"]] + ht_Ref_2@ht_list_param[["width"]]) / 25.4,
            # in inches
            height = (ht_Tum_2@ht_list_param[["height"]] * 1.75) / 25.4  # in inches
        )
        print(final_plot)
        dev.off()
    }
}
